{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pylab as plt\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import os\n",
    "#os.chdir('/Users/ndnguyen20/gtfparse/')\n",
    "from gtfparse import read_gtf\n",
    "#os.chdir('/Users/ndnguyen20/Box/BCM_SMART/')\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "import seaborn as sns\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import multiprocessing\n",
    "from functools import partial\n",
    "#import multiprocessing.Pool as pool\n",
    "#multiprocessing.Pool()._processes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script tried to understand the capture coverage of gene transcripts across human genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_status', 'gene_name', 'level', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_status', 'transcript_name', 'tag', 'transcript_support_level', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>basic</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>basic</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>basic</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>basic</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end  score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1    chr1  HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2    chr1  HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3    chr1  HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4    chr1  HAVANA        exon  13221  14409    NaN      +     0   \n",
       "\n",
       "             gene_id                           gene_type  ...  \\\n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "1  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "2  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "3  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "4  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "\n",
       "  transcript_status transcript_name    tag transcript_support_level  \\\n",
       "0                                                                     \n",
       "1             KNOWN     DDX11L1-002  basic                        1   \n",
       "2             KNOWN     DDX11L1-002  basic                        1   \n",
       "3             KNOWN     DDX11L1-002  basic                        1   \n",
       "4             KNOWN     DDX11L1-002  basic                        1   \n",
       "\n",
       "      havana_transcript exon_number            exon_id ont protein_id ccdsid  \n",
       "0                                                                             \n",
       "1  OTTHUMT00000362751.1                                                       \n",
       "2  OTTHUMT00000362751.1           1  ENSE00002234944.1                        \n",
       "3  OTTHUMT00000362751.1           2  ENSE00003582793.1                        \n",
       "4  OTTHUMT00000362751.1           3  ENSE00002312635.1                        \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#export data from TCGA gencode v22\n",
    "gencode_data = read_gtf('gencode.v22.annotation.gtf')\n",
    "gencode_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#just arrange chromosomes in order\n",
    "chr_names = np.array(['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8',\\\n",
    "       'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',\\\n",
    "       'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22',\\\n",
    "       'chrX', 'chrY', 'chrM'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy and modify from GENCODE_CAPTURE_mismatches\n",
    "\n",
    "class GENCODE():\n",
    "    def __init__(self, chr_name):\n",
    "        self.chr_name = gencode_data[gencode_data.seqname==chr_name]\n",
    "\n",
    "GENCODE_dict_chr={}\n",
    "for chr_name in chr_names:\n",
    "    chr_data = GENCODE(chr_name).chr_name\n",
    "    chr_data.reset_index(drop= True, inplace = True)\n",
    "    GENCODE_dict_chr[chr_name] = chr_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENCODE_genes = {}\n",
    "GENCODE_exons = {}\n",
    "for name in chr_names:\n",
    "    GENCODE_genes[name] = GENCODE_dict_chr[name][GENCODE_dict_chr[name].feature=='gene']\n",
    "    GENCODE_genes[name].sort_values(by = ['start'], axis = 0, inplace= True)\n",
    "    GENCODE_genes[name].reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    GENCODE_exons[name] = GENCODE_dict_chr[name][GENCODE_dict_chr[name].feature=='exon']\n",
    "    GENCODE_exons[name].sort_values(by = ['start'], axis = 0, inplace= True)\n",
    "    GENCODE_exons[name].reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_protein_coding_gene(gene_data):\n",
    "    \"\"\"return protein coding genes in gencode\"\"\"\n",
    "    gene_data = pd.DataFrame.copy(gene_data)\n",
    "    protein_coding_genes = gene_data[gene_data.gene_type=='protein_coding']\n",
    "    protein_coding_genes.sort_values(by = ['start'], axis = 0, inplace= True)\n",
    "    protein_coding_genes.reset_index(drop= True, inplace = True)\n",
    "    return protein_coding_genes\n",
    "\n",
    "def identify_protein_coding_exon(exon_data):\n",
    "    \"\"\"return exon data\"\"\"\n",
    "    exon_data = pd.DataFrame.copy(exon_data)\n",
    "    protein_coding_exons = exon_data[exon_data.gene_type=='protein_coding']\n",
    "    protein_coding_exons.sort_values(by = ['start'], axis = 0, inplace= True)\n",
    "    protein_coding_exons.reset_index(drop= True, inplace = True)\n",
    "    return protein_coding_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENCODE_protein_coding_genes ={}\n",
    "GENCODE_protein_coding_exons ={}\n",
    "#name = 'chr1'\n",
    "for name in chr_names:\n",
    "    GENCODE_protein_coding_genes[name] = identify_protein_coding_gene(GENCODE_genes[name])\n",
    "    GENCODE_protein_coding_exons[name] = identify_protein_coding_exon(GENCODE_exons[name])\n",
    "#making subclasses of genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Use the liftover CAPTURE_hg38 to have the same reference with the GENCODE's\n",
    "capture_data = pd.read_csv('CAPTURE_hg38_no_collapse.txt', header=None, sep =\"\\t\", skiprows=0)\n",
    "class CAPTURE():\n",
    "    def __init__(self, chr_name):\n",
    "        self.chr_name = capture_data[capture_data[0]==chr_name]\n",
    "CAPTURE_dict_chr={}\n",
    "for chr_name in chr_names:\n",
    "    chr_data = CAPTURE(chr_name).chr_name\n",
    "    chr_data.sort_values(chr_data.columns[1], axis = 0, inplace= True)\n",
    "    chr_data.reset_index(drop= True, inplace = True)\n",
    "    CAPTURE_dict_chr[chr_name] = chr_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# in CAPTURE, We later found that some exons lifted over from hg19 to hg38 became alternative haplotypes of a single exon.\n",
    "def detect_collapse_hg38_exons(CAPTURE_chr_data):\n",
    "    #INPUT ORDERED FILES\n",
    "    CAPTURE_chr_data = pd.DataFrame.copy(CAPTURE_chr_data)\n",
    "    for i in np.array(CAPTURE_chr_data.index[1:]):\n",
    "        if CAPTURE_chr_data.loc[i][1] <= CAPTURE_chr_data.loc[i-1][2]:\n",
    "            CAPTURE_chr_data.loc[i][1] = np.min([CAPTURE_chr_data.loc[i-1][1], CAPTURE_chr_data.loc[i][1]])\n",
    "            CAPTURE_chr_data.loc[i][2] = np.max([CAPTURE_chr_data.loc[i-1][2], CAPTURE_chr_data.loc[i][2]])\n",
    "            CAPTURE_chr_data.drop(index = i-1, inplace= True)\n",
    "    CAPTURE_chr_data.reset_index(drop= True, inplace= True)\n",
    "    return CAPTURE_chr_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_capture = pd.DataFrame(columns=[0,1,2])\n",
    "for name in chr_names:\n",
    "    CAPTURE_dict_chr[name] = detect_collapse_hg38_exons(CAPTURE_dict_chr[name])\n",
    "    temp_capture = pd.concat([temp_capture,CAPTURE_dict_chr[name]])\n",
    "capture_data = pd.DataFrame.copy(temp_capture)\n",
    "capture_data.reset_index(drop= True, inplace= True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#capture_data.to_csv(path_or_buf='CAPTURE_hg38_no_collapse.txt', index =None, sep='\\t', header= None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_GENCODE_genes_overlap_w_CAPTURE(coding_genes,merged_exons):\n",
    "    #merged exons are 2-d list of capture exons in a chromosome\n",
    "    overlap_check = []\n",
    "    for gene in coding_genes:\n",
    "        #i = 0\n",
    "        closest_exon_vals = np.argwhere(merged_exons[:,0]-gene[0]<0)\n",
    "        if closest_exon_vals.size > 0:\n",
    "            exon_start_index = np.max(closest_exon_vals)\n",
    "        else:\n",
    "            exon_start_index = np.argmin(merged_exons[:,0]-gene[0])\n",
    "            \n",
    "        i = exon_start_index #good because exons non overlapping\n",
    "        while (i<len(merged_exons)) and ((min(gene[1],merged_exons[i][1])-max(gene[0],merged_exons[i][0]))<0):\n",
    "            i+=1\n",
    "        if (i<len(merged_exons)) and ((min(gene[1],merged_exons[i][1])-max(gene[0],merged_exons[i][0]))>0):\n",
    "            overlap_check.append(1)\n",
    "        else:\n",
    "            overlap_check.append(np.nan)\n",
    "    return overlap_check\n",
    "\n",
    "def detect_CAPTURE_exons_overlap_w_GENCODE(exons, merged_exons_per_chr):\n",
    "    overlap_check = []\n",
    "    for exon_i in range(len(exons)):\n",
    "        #i = 0\n",
    "        closest_exon_vals = np.argwhere((merged_exons_per_chr[:,0]-exons[exon_i][1]<0) | (merged_exons_per_chr[:,1]-exons[exon_i][1]<0))\n",
    "        if closest_exon_vals.size > 0:\n",
    "            #print(closest_gene_vals.size)\n",
    "            exon_start_index = np.max(closest_exon_vals)\n",
    "            \n",
    "        else:\n",
    "            exon_start_index = 0\n",
    "        \n",
    "            \n",
    "            #exon_start_index = np.argmin(merged_exons_per_chr[:,0]-exons[exon_i][0])\n",
    "          #  print(gene_start_index)\n",
    "            \n",
    "        i = exon_start_index\n",
    "            #i = 0 #because exons on GENCODE can be overlapping\n",
    "        while (i<len(merged_exons_per_chr)) and ((min(exons[exon_i][1], merged_exons_per_chr[i][1])-max(exons[exon_i][0], merged_exons_per_chr[i][0]))<0):\n",
    "            i+=1\n",
    "        #print((0,i))\n",
    "        #print(i==len(merged_genes_per_chr))\n",
    "        if (i<len(merged_exons_per_chr)) and ((min(exons[exon_i][1], merged_exons_per_chr[i][1])-max(exons[exon_i][0], merged_exons_per_chr[i][0]))>0):\n",
    "            overlap_check.append(1)\n",
    "        else:\n",
    "            overlap_check.append(np.nan)\n",
    "        #else:\n",
    "            #overlap_check.append(np.nan)\n",
    "    return overlap_check\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this step, try to find how many percent of GENCODE's protein-coding genes mapped to CAPTURE's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENCODE_gene_ovl_CAPTURE = {}\n",
    "percent_overlap = []\n",
    "num_overlap_genes = 0\n",
    "#name='chr1'\n",
    "for name in chr_names:\n",
    "    GENCODE_gene_ovl_CAPTURE[name] = GENCODE_protein_coding_genes[name]\n",
    "    CAPTURE_p_exons = np.array(CAPTURE_dict_chr[name][[1,2]])\n",
    "    GENCODE_gene_list = np.array([GENCODE_protein_coding_genes[name].start,GENCODE_protein_coding_genes[name].end]).T\n",
    "    match_GvsC = detect_GENCODE_genes_overlap_w_CAPTURE(GENCODE_gene_list,CAPTURE_p_exons)\n",
    "    GENCODE_gene_ovl_CAPTURE[name]['ovl?']=match_GvsC\n",
    "    percent = np.nansum(match_GvsC)/GENCODE_gene_list.shape[0]\n",
    "    num_overlap_genes += np.nansum(match_GvsC)\n",
    "    percent_overlap.append(percent)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.9814724524622136,\n",
       " 0.9855189058728882,\n",
       " 0.9925093632958801,\n",
       " 0.9892761394101877,\n",
       " 0.9896432681242808,\n",
       " 0.9961464354527938,\n",
       " 0.9866071428571429,\n",
       " 0.9940209267563528,\n",
       " 0.9908854166666666,\n",
       " 0.990450204638472,\n",
       " 0.9945652173913043,\n",
       " 0.9980582524271845,\n",
       " 0.9875389408099688,\n",
       " 0.9870550161812298,\n",
       " 0.9814502529510961,\n",
       " 0.9894982497082847,\n",
       " 0.9865092748735245,\n",
       " 0.992619926199262,\n",
       " 0.9945167923235092,\n",
       " 0.9962962962962963,\n",
       " 0.9353448275862069,\n",
       " 0.9908883826879271,\n",
       " 0.9914004914004914,\n",
       " 0.8888888888888888,\n",
       " 1.0]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "percent_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9889976784092056"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_genes = np.sum([GENCODE_protein_coding_genes[name].shape[0] for name in chr_names])\n",
    "num_overlap_genes/total_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19596.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_overlap_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step, try to find how many percent of CAPTURE's exons mapped to GENCODE's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9659971975712284\n",
      "0.9629334640362701\n",
      "0.9740166865315852\n",
      "0.9643028846153846\n",
      "0.9621897965704548\n",
      "0.9572247814233337\n",
      "0.9336803547329863\n",
      "0.9421599029257112\n",
      "0.9522188307114346\n",
      "0.9489726027397261\n",
      "0.9597404757943769\n",
      "0.9615510471204188\n",
      "0.9532835024025628\n",
      "0.9045119182746879\n",
      "0.9314516129032258\n",
      "0.9429526582551494\n",
      "0.9548841398509784\n",
      "0.9625502940266172\n",
      "0.9634330867518494\n",
      "0.9629415170816444\n",
      "0.9376130198915009\n",
      "0.9252196271694879\n",
      "0.9493635860542335\n",
      "0.8518518518518519\n",
      "1.0\n"
     ]
    }
   ],
   "source": [
    "CAPTURE_exons_ovl_GENCODE = {}\n",
    "percent_exons_overlap = []\n",
    "num_overlap_exons = 0\n",
    "#name = 'chr13'\n",
    "for name in chr_names:\n",
    "    CAPTURE_exons_ovl_GENCODE[name] = CAPTURE_dict_chr[name]\n",
    "    GENCODE_p_exons = np.array([GENCODE_protein_coding_exons[name].start, GENCODE_protein_coding_exons[name].end]).T\n",
    "    CAPTURE_p_exons = np.array(CAPTURE_dict_chr[name][[1,2]])\n",
    "    match_CvsG = detect_CAPTURE_exons_overlap_w_GENCODE(CAPTURE_p_exons, GENCODE_p_exons)\n",
    "    CAPTURE_exons_ovl_GENCODE[name]['ovl?'] = match_CvsG\n",
    "    percent = np.nansum(match_CvsG)/CAPTURE_p_exons.shape[0]\n",
    "    print(percent)\n",
    "    num_overlap_exons += np.nansum(match_CvsG)\n",
    "    percent_exons_overlap.append(percent)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9537495018636161"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_exons = capture_data.shape[0]\n",
    "num_overlap_exons/total_exons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### total_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "203430.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_overlap_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(213295, 3)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "capture_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "CAPTURE_exon_non_ovl_GENCODE={}\n",
    "for name in chr_names:\n",
    "    CAPTURE_exon_non_ovl_GENCODE[name] =CAPTURE_exons_ovl_GENCODE[name][CAPTURE_exons_ovl_GENCODE[name]['ovl?']!=1]\n",
    "compile_CAPTURE_exon_non_ovl_GENCODE = pd.concat(CAPTURE_exon_non_ovl_GENCODE.values(), ignore_index=True)\n",
    "compile_CAPTURE_exon_non_ovl_GENCODE.to_csv(path_or_buf='CAPTURE_exons_non_ovl_GENCODE.txt', index=None, sep='\\t', mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.046250498136383884"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compile_CAPTURE_exon_non_ovl_GENCODE.shape[0]/total_exons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Try to find how many GENCODE's protein-coding exons on the CAPTURE list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Copy and modify from GENCODE_CAPTURE_matching.ipynb\n",
    "def percent_difference(reference, range_ovl):\n",
    "    rd = (min(reference[1], range_ovl[1])-max(reference[0], range_ovl[0])+1)/(reference[1]-reference[0]+1)\n",
    "    return rd\n",
    "def range_difference(range1, range2):\n",
    "    \"\"\"find overlap regions\"\"\"\n",
    "    rd = min(range1[1], range2[1])-max(range1[0], range2[0])+1\n",
    "    return rd\n",
    "def search_overlap_partners(chr_ref, chr_search):\n",
    "    \"\"\"searching every exon overlapping from one dataset to another\"\"\"\n",
    "    p_ref_matches = []\n",
    "    p_search_matches = []\n",
    "    ref_indexes = []\n",
    "    search_indexes = []\n",
    "    count = 0\n",
    "    percents_ovl = [] \n",
    "\n",
    "    for i in range(len(chr_search)):\n",
    "        \n",
    "        p_search = chr_search[i]\n",
    "        closest_ref_val = np.argwhere(chr_ref[:,0]-p_search[0]<0)\n",
    "        if closest_ref_val.size > 0:\n",
    "            ref_i = np.max(closest_ref_val) #lower start of ref closest to start of search \n",
    "            p_ref = chr_ref[ref_i]\n",
    "        else:\n",
    "            ref_i = 0 #just to make sure\n",
    "            p_ref = chr_ref[ref_i]\n",
    "            #not to let the ref pass over:\n",
    "        while ((max(p_ref[0], p_search[0])-min(p_ref[1], p_search[1]))>0) and (p_ref[0]<p_search[0]) and (ref_i < len(chr_ref)):\n",
    "\n",
    "\n",
    "            ref_i +=1\n",
    "            if ref_i <len(chr_ref):\n",
    "                p_ref = chr_ref[ref_i]\n",
    "                #else\n",
    "        else:\n",
    "            if ((min(p_ref[1], p_search[1])-max(p_ref[0], p_search[0]))>0):\n",
    "\n",
    "                percents_ovl.append(percent_difference(p_search, p_ref)) #percent of search being overlapped\n",
    "                search_indexes.append(i)\n",
    "\n",
    "    return search_indexes, percents_ovl #only good for p_chr_search because chr_search is only called once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "percents_Gexon_on_CAPTURE = []\n",
    "num_Gexons_counted = 0\n",
    "GENCODE_exons_on_CAPTURE ={}\n",
    "GENCODE_exons_out_CAPTURE = {}\n",
    "\n",
    "for name in chr_names:\n",
    "    list_GENCODE = np.array([GENCODE_protein_coding_exons[name].start,GENCODE_protein_coding_exons[name].end]).T\n",
    "    list_CAPTURE = np.array(CAPTURE_dict_chr[name][[1,2]]) \n",
    "    GENCODEwCAPTURE_ex_indexes, percents_ovl= search_overlap_partners(list_CAPTURE, list_GENCODE)\n",
    "    percents_Gexon_on_CAPTURE.append(len(GENCODEwCAPTURE_ex_indexes)/len(list_GENCODE))\n",
    "    GENCODE_exons_on_CAPTURE[name] = GENCODE_protein_coding_exons[name].loc[GENCODEwCAPTURE_ex_indexes]\n",
    "    GENCODEoutCAPTURE_ex_indexes = list(set([*range(len(list_GENCODE))])-set(GENCODEwCAPTURE_ex_indexes))\n",
    "    GENCODE_exons_out_CAPTURE[name]=GENCODE_protein_coding_exons[name].loc[GENCODEoutCAPTURE_ex_indexes]\n",
    "    GENCODE_exons_out_CAPTURE[name].sort_values(by = GENCODE_exons_out_CAPTURE[name].columns[3], axis = 0, inplace= True)\n",
    "    GENCODE_exons_out_CAPTURE[name].reset_index(drop= True, inplace = True)\n",
    "    num_Gexons_counted+=len(GENCODEwCAPTURE_ex_indexes)\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9130223779476406"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_Gexons_counted/np.sum([GENCODE_protein_coding_exons[name].shape[0] for name in chr_names])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_GENCODE_ovl_exons = pd.concat(GENCODE_exons_on_CAPTURE.values(), ignore_index=True)\n",
    "full_GENCODE_non_ovl_exons = pd.concat(GENCODE_exons_out_CAPTURE.values(), ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_GENCODE_ovl_exons.to_csv(path_or_buf='GENCODE_pr-coding_exons_ovl_CAPTURE.txt', index=None, sep=' ')\n",
    "full_GENCODE_non_ovl_exons.to_csv(path_or_buf='GENCODE_pr-coding_exons_non_ovl_CAPTURE.txt', index=None, sep=' ')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Coding lengths of GENCODE's transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_transcript= gencode_data[(gencode_data.feature=='transcript') | (gencode_data.feature == 'exon')] #with exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_protein_coding_transcript = gencode_transcript[gencode_transcript.gene_type=='protein_coding']\n",
    "gencode_protein_coding_transcript.sort_index(inplace=True)\n",
    "gencode_protein_coding_transcript.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcript_length_cal(transcript_data):\n",
    "    #INPUT ORDERED file\n",
    "    #transcript data include transcripts and exons\n",
    "    trans_only_data = transcript_data[transcript_data.feature=='transcript']\n",
    "    trans_only_indexes = np.array(trans_only_data.index)\n",
    "    list_lengths = np.array(transcript_data.end - transcript_data.start)+1\n",
    "    trans_lengths = np.empty([trans_only_data.shape[0],])\n",
    "    nums_exons_per_trans = np.empty([trans_only_data.shape[0],])\n",
    "    for i in range(len(trans_only_indexes)-1):\n",
    "        trans_length = np.sum(list_lengths[trans_only_indexes[i]+1:trans_only_indexes[i+1]])\n",
    "        trans_lengths[i] = trans_length\n",
    "        nums_exons_per_trans[i] = trans_only_indexes[i+1]-trans_only_indexes[i]-1\n",
    "    \n",
    "    #final transcript:\n",
    "    trans_lengths[-1]= np.sum(list_lengths[trans_only_indexes[len(trans_lengths)-1]:-1])\n",
    "    print(len(trans_lengths))\n",
    "    nums_exons_per_trans[-1] = transcript_data.shape[0]-trans_only_indexes[len(trans_lengths)-1] -1\n",
    "    trans_only_data['length'] = trans_lengths\n",
    "    return trans_only_data, nums_exons_per_trans\n",
    "\n",
    "def df_edit_col_by_label(df,label, lable_col_name, col_name, val_2_col):\n",
    "\n",
    "    df_w_label = df[df[lable_col_name]==label]\n",
    "    save_index = np.array(df_w_label.index)\n",
    "    df.loc[save_index,col_name] = val_2_col\n",
    "    return df\n",
    "    \n",
    "def percent_transcript_covered_CAPTURE(chr_names, CAPTURE_dict, transcript_data):\n",
    "    #need to group CAPTURE dict into chromosome categories\n",
    "    trans_only_data, no_use = transcript_length_cal(transcript_data)\n",
    "    trans_only_indexes = np.array(trans_only_data.index)\n",
    "    #keep index no reset\n",
    "    trans_ovl_lengths = np.empty([trans_only_data.shape[0],])\n",
    "    exons_only_data = transcript_data[transcript_data.feature=='exon']\n",
    "    reset_index_groups = []\n",
    "    exon_length_groups = []\n",
    "    for name in chr_names: \n",
    "        exon_ovl_indexes, exon_ovl_percents = search_overlap_partners\\\n",
    "        (np.array(CAPTURE_dict[name][[1,2]]), \\\n",
    "        np.array(exons_only_data[exons_only_data.seqname==name][['start', 'end']]))\n",
    "        #because the func return ordered indexes\n",
    "        reset_exon_ovl_indexes = np.array(exons_only_data[exons_only_data.seqname==name].index)[exon_ovl_indexes]\n",
    "        reset_index_groups.append(reset_exon_ovl_indexes)\n",
    "        \n",
    "        exons_only_ovl_data = exons_only_data[exons_only_data.seqname==name].loc[reset_exon_ovl_indexes]\n",
    "        #because percents over the exon length\n",
    "        exon_ovl_lengths = np.array(exon_ovl_percents)*np.array(exons_only_ovl_data.end-exons_only_ovl_data.start+1)\n",
    "        exon_length_groups.append(exon_ovl_lengths)\n",
    "        \n",
    "        #MAKE GROUPS OF RESET INDEXES AND GROUPS OF EXON_OVL_LENGTHS\n",
    "        exons_only_data['ovl_length'] = np.zeros([exons_only_data.shape[0],])\n",
    "    for i in range(len(reset_index_groups)):\n",
    "        name = chr_names[i]\n",
    "        exons_only_data[exons_only_data.seqname==name] = \\\n",
    "        exons_only_data[exons_only_data.seqname==name].loc[reset_index_groups[i]] #after this, NaN at other loc\n",
    "        exons_only_data[exons_only_data.seqname==name].dropna(how='all', inplace=True)\n",
    "        exons_only_data = df_edit_col_by_label(exons_only_data, name, 'seqname','ovl_length',exon_length_groups[i])\n",
    "        #exons_only_data[exons_only_data.seqname==name]['ovl_length'] = exon_length_groups[i]\n",
    "        \n",
    "        \n",
    "    for i in range(len(trans_only_data)-1):\n",
    "        #index_range = [exon_indexes]\n",
    "        trans_ovl_length = np.sum(exons_only_data['ovl_length'] \\\n",
    "                                  .loc[(exons_only_data.index>trans_only_indexes[i]) & \\\n",
    "                                       (exons_only_data.index<trans_only_indexes[i+1])])\n",
    "        trans_ovl_lengths[i] = trans_ovl_length\n",
    "    \n",
    "    trans_ovl_lengths[-1] = np.sum(exons_only_data['ovl_length'] \\\n",
    "                                  .loc[(exons_only_data.index>trans_only_indexes[-1])])\n",
    "    trans_only_data['%_exon_ovl'] = trans_ovl_lengths/trans_only_data.length\n",
    "    return trans_only_data\n",
    "                            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "143615\n"
     ]
    }
   ],
   "source": [
    "transcript_ovl_result = \\\n",
    "percent_transcript_covered_CAPTURE(chr_names, CAPTURE_dict_chr, gencode_protein_coding_transcript)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate transcripts/gene; exons/transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "143615\n"
     ]
    }
   ],
   "source": [
    "trans_only_data, nums_exons_per_trans = transcript_length_cal(gencode_protein_coding_transcript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_only_data['num_exons'] = nums_exons_per_trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, 'number of exons per transcript')"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkgAAAFSCAYAAAAJuyDPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3XlYVeX+//8XQygOOZs/U8xpY4oCKpk5JWqO6C8tNXNMKzXNISstNcs8HbWyjEo5Dg2a5+jHIcsGc8iBc04nNacSO0UClUOCmBMicH//8Owd+wZlYwwKz8d1cV17r3Wve9/3m8XitddaG7yMMUYAAABw8S7sAQAAAFxvCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYfK91w1Onzikjw+TlWAAAAPKUt7eXKlQonevtrjkgZWQYAhIAACiSuMQGAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAW38IeQE4++OA9JSTESZJOn06WJJUrV961vmbNWhowYHChjA0AABRN131ASkiI0+H//iCfkuWVnnI5IP32e5okuZ4DAADkpRviEptPyfIqVauDfEqWdz12PgcAAMhrN0RAAgAAKEgEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADA4lvYA7iS6Ojted5Xq1Zt86xPAABQdF23AWnnzm153hcBCQAAeIJLbAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFt/CHkBBeuihAYU9hD/N37+ULlw471FbX19fpaWlSZKCg5tp377drnUVKlTQqVOnJElVqlTVb7+dcK2rWrWqTpy4/LxSpUpKTEyUJI0a9biOHz+uNWv+IV9fX/Xq1UerV//DtV3btu20ffs2SVKrVncrOvpL1zqH43Z9//0hSVKHDvdo8+aNGjJkuNq166D4+CP6y19mKDU1Vffd11+rV/9DXbv21CeffKghQx7Sjh3bJUljx05UTMx3WrgwUjfd5KepU59XzZq1JMmtDz8/Pz3++BNatuwdHTt2VPfd11/r169RamqqJk16Rg0bBl21bvHxRzR79kx169ZLq1f/XaNGPa769QO1YMEbGjXqcZUrV96j+v8ZzjFMnjzdNUdP2pYte7MiI+cpLS1Nvr6+Gjt2YoGM91plHrsxRrNnz9To0RP00UdrPKp1cvIpvfbaXJ04cUxTpjyXY62AgpacfEoLFryhiIh79dZbr3n0M12cOOtTUMfW3PCZMWPGjGvZ8MKFVBmTx6PJJDp6u+tx0u/ndVP5Orp0+idJ0k3l60iSLp3+SRXLlVLr1u086isx8WQ+jbbgpKVd8rhtRkaG6/Hx40fd1qWkpLgenz9/zm3duXN/PL9w4YLr8Tff7NZ33x1w9X3o0Ldu28XFxbkeJyQccVuXufY//fSjJGnfvr3q1auP5s6dpeTkZEnSd98dlCT997+HXW1OnUrSqVNJSk29qLVr/0/GZCgjI13ffx+j8PB7JMmtj/T0dO3d+41OnUpy9Zmeni5J2rt3j7p162mXys3cubN06lSSDh066Jr3xYsp2rPna128eFHBwaFX3T4vOMeQeY6etE1MPKlvvtml06eTXTUriPFeq8xj3737a506laR9+/bo+PFjHtV61aoV2rdvj9LS0jyqFVDQVq1aoT17vtbevXt04cJ59lOLsz75eWz18vJSqVJ+ud6u2FxiO3z4UGEP4YaXnp6Wxz0arVmzUr/++stV2zht27bFbQy//vqLEhLiFB9/JEsfdujLvNwZwrKTXV/p6Wnatm2LjDHauXO7Tp9Ovsp4/7zMY3DO0dO227dvdVu/Y8e2fB/vtbLH7nx8/vw5j2qdnHzKbb451QooaMnJp7Rz5zYZY1zHJPbTP2SuT0EcW3Pruj2D9Mkn63XixHGdPn1al9Iz5FfRkeUM0sXEQzp35pQOHtyn6OjtV/yKj4/L1ZkXFJzvv4/xuK3JZodznnk4c+aMx/1c7SzS3Lmzsu3L+dpeXl5KSbmQr2dl7DFc7R2n3dauUUaGuW7PIl2p1k451XrVqhWKjf3RbRnvznE9WbVqhY4cOSJjMtyWs59elrk++Xls5QwSiqXMZx48daWzS87+riY9PU3/+ld0rl4vt+wxXG1MOc/d5Pt4r9WfrXV263K7LwD56V//is72zDv76WWZ61MQx9bcum5v0i5Xrrzrhq0fErK/d8jbt6QCalbW009Pu2pfs2fP5BJbEVW9+q2ScnfAKVWq9FX7u1pfPj6+atmylecDvAb2GJxz9KRtVl75Pt5r9Wdr3bJlK23duilLn8D1omXLVtq+/cssIYn99LLM9SmIY2tucQYJhapHj//f47be3ll310cfHaNHHnksV685evS4K667Ul/O1/b29lbPnr1z9Xq5ZY/h0UfHeNzWx8fH7bmvr2++j/da5fR9y6nWERH3Zpnv1WoFFLSIiHvl7e2VZTn76WWZ61MQx9bcKjYBKTDw9sIewg3PxyevTzh6qXfvvjm8m/rj4NKuXbjbGKpXv1U1a9ZSQMBtWfq40lmiUqVKX/Vj/tn15ePjq3btwuXl5aXWrdvm+0dRM4/BOUdP27Zt295tfZs27a67j8462WN3Pi5VqrRHtS5fvoLbfHOqFVDQypevoNat2/3vHpjLxyT20z9krk9BHFtzq9gEpKLC37+Ux219ff8IE8HBzdzWVahQwfW4SpWqbuuqVv3jeaVKlVyPH3lktHr37ufqu0+ffm7btW37x59baNXqbrd1DscfAbVDh8s3Jw4Z8tD/+n1Mfn6Xb6C7777+8vLyUrduvf7XZpjq1KmnOnXqqWfP3hoxYqQk6aab/NzehWXuw8/PT6NHP65q1f4/V5/OdVc7e5S5L39/f/Xp098174iIe1W/fmCBvcNxjsGTd5qZ20ZE3Ks6deopIOA2V82uZ5nH7nw8atQ4j2sdEXGvAgJuU8mSJXlXjuuS89gxatTjHv9MFycFfWzNDS+T3UeDPJCYeFYZGfn3MbbZs2e6Hv+QcFKlanXQ+bjNkqRStTpIks7HbVY9D+9BkpRjOwAAULR4e3upUqUyud8uH8YCAABwQyMgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAxbewB3AlrVu3kyTt3Lktz/oCAADwxHUbkFq1aispbwKSsy8AAABPcIkNAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACy+hT0AT6SnJOt83GalpyRLks7HbXYtlyoX4sgAAEBRdN0HpJo1a7kenz59ebjlypX/35LKbusBAADygpcxxlzLhomJZ5WRcU2bAgAAFAhvby9VqlQm99vlw1gAAABuaAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsvte6obe3V16OAwAAIM9da17xMsaYPB4LAADADY1LbAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAACW6yogffzxx+revbuaNGmirl27at26dYU9pAKXlpamJk2aKDAw0O0rNDTU1Wbnzp3q06ePgoODFR4eriVLlhTiiPPXoUOH1KhRIx07dsxtuSc1OHDggAYNGqTQ0FC1bt1ar776qi5dulRQQ89XV6pLp06dsuw7gYGBSkpKcrUpanXJyMjQihUrFBERodDQUHXs2FEvvfSSzp4962rjyZyPHDmikSNHqnnz5mrRooWee+45tz5uNJ7UZejQodnuLwcOHHC1KWp1kSRjjN555x117txZTZo0Uc+ePfXRRx+5tSmOxxhP6lKcjjHX/L/Y8tqnn36qSZMmafDgwWrTpo02bdqkp59+WiVLllSXLl0Ke3gF5qefftLFixc1e/Zs3Xbbba7l3t6Xs+yePXs0cuRIde3aVePGjdPu3bs1Z84cGWM0fPjwQhp1/oiNjdWjjz6qtLQ0t+We1CAuLk5Dhw5VaGioXnvtNf3444+aN2+ezp49q+nTpxfGdPLMlepy7tw5JSQk6IknntAdd9zhtu7mm2+WVDTrsmjRIr322msaPny4WrZsqZ9++knz58/XDz/8oMWLF3s059OnT2vIkCGqUqWKZs+ercTERM2dO1fHjh3TwoULC3mG1yanukhSTEyMBg8erO7du7ttW7duXUlFsy6StHDhQs2fP19jx45VSEiItm/frkmTJsnHx0fdunUrtseYnOpS7I4x5jrRsWNHM378eLdl48aNM126dCmkERWO9evXmwYNGpjz589nu37IkCHm/vvvd1s2Z84c07x5c3Px4sWCGGK+u3Tpklm2bJkJDQ01d9xxh3E4HObo0aOu9Z7U4JlnnjHt2rVzq8ny5cvN7bffbo4dO1YwE8ljOdVl9+7dxuFwmB9++OGKfRS1umRkZJiwsDAzY8YMt+UbNmwwDofDfPfddx7N+c033zQhISEmKSnJ1ebLL780DofD7N27t2Amk4c8qcuxY8eMw+Ew27Ztu2I/Ra0uxhiTmppqwsLCzAsvvOC2fODAgeaBBx4wxhTPY4wndSlux5jr4hJbQkKC4uPjdc8997gt79y5s2JjY5WQkFBIIyt4hw4dUkBAgPz9/bOsu3jxonbt2pVtnX7//Xft2bOnoIaZr3bv3q2XX35ZDz30kCZNmuS2ztMaREdHq3379vLz83O16dKli9LT07Vz5878n0Q+uFpdpMv7TokSJdzOPNqKWl3OnTunnj17qkePHm7L69SpI0mKj4/3aM7R0dEKCwtThQoVXG1at26t0qVLa9u2bQUwk7zlSV1iYmIkSYGBgVfsp6jVRZJ8fHz0/vvv65FHHnFbftNNN+nixYvF9hiTU12k4neMuS4CUmxsrCSpdu3abstr1aol6fJlp+Li8OHD8vPz0/DhwxUaGqqwsDBNnz5dZ8+eVUJCgi5dulTk61S3bl1t2rRJY8aMkY+Pj9s6T2pw4cIFHT16NEubihUrqkyZMjdsna5WF+nyvlO+fHlNnDhRzZs3V2hoqCZMmKDffvtNkopkXcqUKaOpU6eqWbNmbss3bdok6XLNPJlzbGxsljY+Pj6qUaNGkaxLvXr1FBMTIz8/P82fP18tWrRQ48aN9fDDD7vNt6jVRbp8u0JgYKBuueUWGWN08uRJRUVF6Z///Kf69etXbI8xOdVFKn7HmOsiIJ05c0bS5R/qzEqXLi1JN/wNgbkRExOj+Ph4tWvXTlFRURo9erQ+/vhjjRo1qtjUqXLlyqpUqVK26zypwZXaONvdqHW6Wl2ky/vOyZMnVb9+fS1YsEBTpkzR119/rcGDByslJaXI1sW2b98+RUVFqWPHjq77InKa85kzZ4pVXerWrauYmBilpqaqZMmSioyM1KxZsxQfH68HH3zQ9QuvqNdl48aNatWqlV555RW1a9dOPXv2LNbHGKfs6iIVv2PMdXGTtjFGkuTl5ZXtcucNysXBvHnzVK5cOddp77CwMFWqVElPPvmkoqOjJWWtk1NxqNOV9hUnb2/vq7YxxhTZOk2dOlXGGAUHB0uSmjdvrrp162rAgAFav3692rVrJ6lo12X37t0aOXKkatSooRdffFGpqamSPJtzcaqLJI0aNUr9+vXTnXfe6WoXGhqqrl27atmyZZowYYKkol2Xhg0batmyZTp8+LBef/11PfLIIxo/fryk4n2Mya4u7733XrE7xlwXAals2bKSsp4BOXfunNv64sD+ZIAk3X333W7P7To5nxeHOl1pX8lcA+e7l+zerZw/f77I1qlJkyZZljVr1kxly5ZVTEyM65NKRbUun3zyiSZPnqzbbrtNixYtUoUKFVzHkJzmXKZMmWzbnDt3Trfeemv+DjyfZVcXSXI4HFna1qxZ03V2SSradZEuz7dmzZoKCwtTmTJl9PTTT7vCT3E+xmRXl2+++cbtz804FeVjzHUR55zXK+Pj492Wx8XFua0v6hITE7Vq1aosN6WnpKRIkipVqiQfH58sdXI+Lw51CggIyLEGpUuX1i233OLaf5wSExN19uzZIlmn8+fPa/Xq1a5fbE7GGF26dEkVKlQo0nVZunSpJk6cqJCQEC1fvlxVq1aVJI/nXLt27Sxt0tPT9fPPPxfJuhhjtG7dOu3atSvLNikpKa4QVRTrkpycrHXr1un48eNuyxs2bChJ+vnnn4vlMSanuhw5cqTYHWOui4BUq1Yt1ahRQ5999pnb8o0bN+q2225T9erVC2lkBcvLy0vTp0/XsmXL3JZ/8skn8vHx0V133aXmzZtr48aNrnc5kvT555+rbNmyCgoKKughF7gSJUp4VINWrVpp69atrksszjY+Pj7ZnqW70ZUoUUKzZ89WZGSk2/LNmzcrJSXFNeeiWJdVq1bpr3/9q7p27apFixZleZfqyZxbtWqlr776SsnJya42O3fu1Pnz53XXXXcVzETy2NXq4uXlpcWLF+svf/mLMjIyXMu//fZbxcfHF+m6ZGRkaPLkyfrHP/7httx5C0Pjxo2L5TEmp7oEBwcXu2OMz4wZM2YU9iCky6ct3377bZ06dUpeXl5aunSp1q5dq+eee07169cv7OEVCH9/fyUnJ2v58uXKyMhQRkaGPvzwQ82fP18DBgxQRESEqlWrpgULFujHH3+Uv7+/1q1bp7/97W8aO3asWrRoUdhTyHOHDh3S5s2bNWzYMNdpbU9qULt2bS1ZskS7du1SuXLl9OWXX2ru3Lm6//77FRERUZhTyhN2Xby9veXr66v33ntPp0+flq+vrzZv3qxZs2apTZs2evTRRyUVvbokJiZqxIgRuuWWW/TEE08oMTFRx44dc335+fmpYcOGOc65Xr16+vvf/65NmzapUqVK2rNnj2bMmKEWLVq4ancj8aQuNWvW1LvvvqsjR46oTJky+uqrrzR16lTddtttmjZtmry9vYtcXaTLx9mkpCS999578vX1VWpqqj788ENFRkaqd+/e6tOnT7E8xnhSl2J3jCnAv7mUoxUrVphOnTqZoKAg07VrV7N27drCHlKBS01NNVFRUaZz584mKCjIdOjQwSxcuNCkp6e72mzcuNH06NHDNGrUyISHh5vFixcX4ojz1+rVq7P8QURjPKvB119/be6//34TFBRk2rRpY1555RWTmppaUEPPV1eqy8qVK02PHj1MkyZNTJs2bcycOXPMhQsX3NoUpbqsXbvWOByOK36tW7fOGOPZnA8fPmyGDBlimjRpYlq2bGmmTZtmzpw5UxjT+tM8rcsXX3xh+vTpY0JCQsydd95ppk2bZk6dOuXWV1Gqi5PzOHvPPfeYoKAg07Fjx2s6zhalnyVjPKtLcTrGeBmT6RwiAAAAro97kAAAAK4nBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAq5BYGCgJk+eXNjDuKrU1FRNmTJFTZs2VdOmTbVly5bCHlKxkZiYqPPnzxf2MPLcG2+8ocDAQP3888+53jYjI+OatgMKCwEJKKJWrlypNWvWqGPHjpoyZUqx+F9914Nt27apS5cuSkpKKuyh5LlOnTppzpw5qlixYq62O3v2rPr27au1a9fm08iAvOdb2AMAkD8OHz4sSZo+fbrr/9gh/+3fv1+///57YQ8jXzRo0EANGjTI9XbJyck6cOCA2rVrlw+jAvIHZ5CAIurSpUuSRDgCgGtAQMINIzw8XNOnT9eHH36o7t27q3Hjxrrnnnu0fPnyLO0GDRqU7faZl4eHh+uFF17QqlWr1LlzZzVp0kR9+vTR/v379dtvv2ncuHEKDQ1VmzZtNG/ePGVkZGTpc8GCBWrTpo2Cg4M1ePBg7d+/P0ubrVu3qn///goODlZYWJjGjh2rn376ya1NYGCgXnvtNY0cOVJBQUHq1q2b0tLSrliLTZs2qX///mrSpImaN2+ukSNHKiYmxq0/5+WMwMDAbOuRmzGuWLFCgYGBmjNnjtt2w4cPV1BQkNtr5zQ255iioqK0dOlSdezYUUFBQYqIiNCnn37q1u7XX3/V2LFj1bp1azVu3FjdunXT3/72t2y/F5mFh4fr2Wef1apVq9ShQweFhISof//++ve//52l7TfffKNhw4YpNDRUoaGheuihh7J8H8PDwzV16lQ988wzaty4sdq2bZvtJbTJkycrMjJSktShQwdX3QcNGqThw4dr3rx5Cg0NVcuWLV1n+D777DMNHDhQzZo1U1BQkMLDwzVnzhylpqa6+nVuv337dvXu3VuNGzfW3XffrTfeeMOtFqmpqZo1a5Y6dOigoKAgtWvXTs8//7xOnz7tNs7jx4/rmWeeUevWrRUaGqo+ffpo06ZNbvPo0qWLli9frrCwMIWFhWn79u1Z7kF644031LBhQ8XGxmrQoEEKDg5WeHi43nrrLaWnp0uSvvrqK3Xo0EGSFBkZec33MAEFjYCEG8qOHTs0a9Ysde7cWVOmTJG/v79eeOEFbdu27Zr627x5s15//XXdd999GjNmjGJjYzV27FgNGzZM3t7emjx5shwOhxYsWKAPP/zQbdvPP/9cS5cuVf/+/fXYY48pNjZWgwcP1n//+19XmzVr1mjUqFHy9/fXk08+qaFDh+qbb75R3759s4Skd999VykpKZo6dar69u0rX9/sr4AvX75cjz32mC5duqSJEydq6NCh2r9/vx544AHXL/Y5c+aoefPmrscjR468Yg08GWP//v3VokULvfvuu/r+++8lXb7HaefOnXr88cddl108GZvTihUr9M4776hv37566qmndP78eU2YMMHV/6VLlzRixAh9++23Gjp0qKZNm6batWvr5ZdfVlRU1NW/sZL++c9/6oUXXlDnzp01btw4JSUlacSIEfrPf/7jahMdHa1BgwbpzJkzGjdunEaNGqVff/1VDz74oHbt2uXW34YNGxQTE6Nnn31Wffv2zfY+nH79+qlTp06SpClTprjVfc+ePdqwYYOefPJJ3XvvvapXr55WrVqlcePGqWzZspo0aZKeeuop3XrrrVq8eHGWOX7//fcaP368WrRooalTp6pmzZqKjIzUihUrXG2cgb979+567rnn1LlzZ61cuVITJkxwtUlOTlbfvn21YcMG9ezZU0899ZRKliypMWPGuIWko0eP6s0339SYMWPUt29fBQcHZ1tnY4yGDRumEiVK6Mknn1SjRo30+uuv6/nnn5ck1a1bV1OmTJF07fcwAYXCADeI9u3bm8DAQHPo0CHXshMnTpjAwEAzceJEt3YDBw7MdvvMy539xcTEuJbNnj3bOBwOM378eNeyc+fOmUaNGrm9hsPhMLfffrvbtkeOHDGNGjUyY8aMMcYYc+bMGdO0aVMzYcIEt3GcOHHChIWFmdGjR7v116xZM3P69Omr1iApKckEBweb++67z1y8eNG1PCEhwbXc6emnnzYOh+Oq/eVmjPHx8SYkJMQMGDDAHD161DRt2tT079/fpKen53qDbQnBAAAKr0lEQVRsDofDhISEmBMnTriW7d271zgcDvPqq68aY4zZt2+fcTgc5tNPP3W1ycjIMA899JB56qmnrjqv9u3bG4fDYb744gvXssTERNO8eXPTt29fY4wx6enppkOHDqZ///4mLS3N1e7cuXOmU6dOplevXm79NWjQwMTFxV31dY0xZv78+cbhcJiEhATXsoEDBxqHw2H+/e9/u7Xt0qWL6devn8nIyHAtu3Tpkmnbtq3p0aNHlu03b97sWpaSkmLCwsJMv379XMuaNGlinn/+ebfXmDdvnundu7c5e/asMcaYOXPmGIfDYXbt2uXWV8eOHU2fPn2MMX/sO6tXr77q3JzPR48e7TaHJ554wgQGBpoffvjBGHN5H3A4HGb+/Pk51g+4XnAGCTeU2rVru90kWqVKFVWuXFknT568pv4CAgIUGBjo1r8k11kASSpVqpQqVaqk3377zW3bNm3auG1bq1YttWnTRjt37lR6erqio6N19uxZdezYUUlJSa4vHx8f3Xnnndq5c6fbZbTg4GDdfPPNVx3vv/71L124cEHDhg2Tn5+fa3mNGjXUs2dP7d+/XydOnPB4/rkZY82aNTV+/Hjt2rVLAwcOVEZGhmbPni1vb+9rGluzZs1UpUoV1/Pbb79dklx1rlq1qry8vLRw4ULt2LFDqamp8vLy0uLFizV79uwc51anTh117NjR9bxixYrq1auX9u3bp8TERH333XdKSEhQx44ddfr0adfcU1JS1L59ex06dEjHjh1zbR8QEKCAgACPa2srWbKkwsLC3JatX79eUVFR8vLyci1LTEzUzTffnOXPBPj7++vuu+92PS9RooRq167ttu9Xq1ZNn3zyidasWeO6UXz8+PFavXq1SpcuLUn68ssv1ahRIzVr1sytr6ioKM2fP9/tNVu3bu3R3B555BG3OQwbNkzGGG3dutWj7YHrEZ9iww0lu1Pzfn5+Od6TciWVKlVye+7j45Pt6/j4+MgY47asTp06WfoLCAjQli1blJSUpPj4eElyu7xhS0pKUtWqVbN9zew4793I7rXr1q0r6fJ9O84+c5LbMQ4aNEjr16/XwYMHNWnSJLfAkNux2fN1hirn97JatWp68skn9eqrr2rEiBEqVaqUWrZsqW7duqlr166u79WV1KtXL8uyWrVqyRijX375xTXeOXPmZLm3yuno0aOqVq2apKz7Sm6VL1/eFSadbrrpJn399df6+OOPFRsbq/j4eCUmJkqSbr311hy3t/f9GTNmaPz48ZoyZYqmTZumkJAQderUSX369FHZsmUlSb/88ovCw8OzjM/55iAzT+fs/P461apVy/VawI2KgIQbiv0LIjecN41mdqX7fDK/G84N5y8rHx8f1+OZM2eqRo0a2bYvV66c63FOv/Bz4gxwN910k8fb5HaMv/32mytUbd68WcOHD/foe5Ld2DzZbvjw4erRo4e++OILbdu2TdHR0dq8ebPWrVunRYsWXXXb7Org3Acyf3/GjRunkJCQbPvIHPb+7Pcnu+1feeUVRUVFqWHDhgoJCVGvXr0UGhqqmTNn6ujRo25tPalXy5YttXXrVtdXdHS0XnrpJb3zzjtas2aNKlasqPT0dI/3b0/nbNc6888BcKMiIKHI8fb2dvsEkCSlpaXp1KlTf+oSiS27d8dxcXEqW7asKlSo4DoDULFiRd11111u7b766itlZGS4XYryhLPP2NjYLH+PJjY2VpJcZzxy05+nY5wxY4ZSU1M1YcIEzZs3T++++66GDRuWL2NLTk5WTEyMmjZtqoEDB2rgwIE6f/68Jk+erM8//1yHDx92u8Rpcwa5zOLi4uTj46MaNWq49pFSpUplmfv+/ft1+vRplSxZ0uPx5tYvv/yiqKgo9erVK8sZrGu5ZJyamqpDhw6pWrVq6t69u7p3766MjAwtXbpUc+bM0YYNGzRo0CBVr14929qsXbtWu3fv1vTp03P92gkJCW5n7I4cOSLpjzNJwI2Ie5BQ5FSuXFk//fSTUlJSXMu2bNmiixcv5unr7NixQ8ePH3c9//7777Vz506Fh4fLy8tLd911l0qUKKFFixa5/iaRdPkj1qNHj9bLL7+c6zNVzj6XLl3qFgKPHTumjz76SE2aNMnVpaDcjPHjjz/Wli1b9Nhjj2nkyJFq2bKlXnvtNcXFxeXL2KKjozVkyBC3f5FSqlQpORwOSTmfnThw4ID27t3ren7y5EmtX79ed955p8qVK6egoCBVqVJF77//vs6dO+dqd/bsWddlqms5A+I802NfkrU5P3pvXwrctm2bjhw5ctU/85Cd5ORk9evXTwsXLnQbS+PGjd3G1bZtWx04cEAHDx50tbt06ZIWL16sgwcP5jq0S9L777/v9nzp0qXy9fV1Xcpz1vFaL4UDhYEzSChyevTooZkzZ2rEiBHq2bOn4uLitHLlyiz3dPxZfn5+GjBggAYNGqQLFy7onXfe0c0336zx48dLunxWZuLEiXrppZfUr18/9ezZU2lpafrggw908eJFPf3007l+zQoVKrj6fOCBBxQREaFz585pxYoVysjI0NSpU3PVn6djTEpK0osvvqj69eu7zhg999xzioiI0LPPPqv3338/z8fWvn171a5dW88++6y+/fZbBQQEKDY2VsuXL9edd96Z7T1Gmfn5+enhhx/WkCFDVLJkSX3wwQfKyMjQU089JenyZaFp06Zp/Pjx6t27t+677z6VKFFCq1at0q+//qqXX375ipdgc6qpJC1atEht27Z1/Q0gW7169VS9enUtWLBAFy9eVLVq1bR//36tXbtWJUqUcAttnqhataoiIiL0wQcf6MKFCwoNDVVycrKWLVumypUrq2vXrpKkRx99VJ999pmGDBmigQMHqmrVqtqwYYN+/PFHLV68ONfzlS6ffTp79qyaNm2qHTt2aOvWrXrsscdcP3PO+6e2bNmi6tWr65577nG7dAtcjwhIKHIGDBig5ORk/d///Z9mzpypBg0aKDIyUkuWLMnTfyDar18/eXl5uX7BtWjRQpMnT1b16tVdbYYOHapbbrlFS5cu1bx581SyZEk1atRIc+fOdfsUUW4MHTpUVatW1ZIlS/Tqq6/K399fd9xxh8aMGXPVS05X6y+nMc6cOVPJycmKjIx03W9Su3ZtPfzww3rrrbf0wQcf6MEHH8zTsZUqVUpLlizR/Pnz9dFHH+nkyZOqUqWKBgwYoDFjxuS4fUhIiLp376633npLZ86cUfPmzfXEE0+4Xf7r3LmzlixZorfffltvvfWWvL29Vb9+fb399ttq3759rsbr1L17d23cuFFr1qzRf/7znysGJD8/P0VFRemvf/2r3nvvPRljFBAQoGeeeUZpaWmaNWuWDh48mKv/oTdz5kzVrFlTGzZs0IYNG+Tv76+WLVtqwoQJruBWuXJlrVy5Uq+88or+/ve/KzU1VQ0aNNCSJUvUsmXLa5pzZGSk3nzzTW3cuFE1a9bUzJkz1bdvX9d6f39/TZgwQYsXL9aLL76ogIAAtWjR4ppeCygoXian88AAcIMJDw/XrbfemuXSD/LWG2+8ocjISG3evPmKN/kDNyruQQIAALAQkAAAACwEJAAAAAv3IAEAAFg4gwQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYPl/+chkrKkfsB4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, axes = plt.subplots(figsize=(10,5))\n",
    "sns.boxplot(nums_exons_per_trans, width=0.5, ax = axes)\n",
    "axes.set_xlabel('number of exons per transcript')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.245608049298472"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.average(nums_exons_per_trans)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Understand number of transcripts per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_gene_ids, gene_indexes = np.unique(trans_only_data.gene_id, return_index=True)\n",
    "#prepared for sorting indexes\n",
    "merged_table = pd.DataFrame(unique_gene_ids, index = gene_indexes)\n",
    "merged_table.sort_index(inplace= True)\n",
    "gene_ids = np.array(merged_table[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "#select GENCODE protein-coding genes on one table\n",
    "gencode_genes = gencode_data[gencode_data.feature=='gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "nums_trans_per_gene = []\n",
    "for id_tag in gene_ids:\n",
    "    num_trans = transcript_ovl_result\\\n",
    "    [transcript_ovl_result.gene_id==id_tag].shape[0]\n",
    "    nums_trans_per_gene.append(num_trans)\n",
    "#gencode_only_protein_coding_genes['num_trans'] = nums_trans_per_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1a6bf32710>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYwAAAEGCAYAAAB2EqL0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3XtcVHXCP/DPXLnMcFdRGAQVNC95SdRMVFovgOaa7tO25ma9Wi/btj1Ptf7ysutTtu2qtKu2Ye6ztND6lG4v99msVNTI1DS74L1U1EAYFKwYBAaYGWbm/P7AmRiYYY4Ic4D5vF8vX+b3nDnznfOi+fC9nO9XJgiCACIiIi/kUleAiIi6BwYGERGJwsAgIiJRGBhERCQKA4OIiERhYBARkSgMDCIiEoWBQUREojAwiIhIFAYGERGJwsAgIiJRGBhERCQKA4OIiERRSl2BjlJVVQe7nQvvEhF5I5fLEBGhue3X9ZjAsNsFBgYRUSdilxQREYnCwCAiIlEYGEREJAoDg4iIRGFgEBGRKAwMIiIShYFBRESi9JjnMDrLtre3o7LqpsfjURHhWLTwER/WiIhIGgwMLyqrbmJMylyPx08dfc+HtSEikg67pIiISBQGBhERicLAICIiURgYREQkCgODiIhEER0Yu3fvxuzZszFy5EhkZGRg165dot9kw4YNePzxx1uVFxQUYMiQIa3+LFu2TPS1iYjIN0RNq83Ly8Py5cuxaNEiTJ48Gfn5+VixYgUCAwORnp7e5mvfeust5OTkYOLEia2OFRYWIjg4GLm5uS7loaGht/ERiIjIF0QFxsaNG5GRkYHVq1cDACZPnozq6mq8+uqrHgPjxo0byMzMxN69exESEuL2nIsXLyIpKQmjR49uZ/WJiMhXvHZJ6fV6lJaWYubMmS7laWlpKCoqgl6vd/u6TZs24fz588jNzcXQoUPdnnPhwgUMGTKkHdUmIiJf8xoYRUVFAIABAwa4lMfHxwMAiouL3b5u8eLF2LNnD+699163x+12Oy5fvoyKigrMmzcPI0aMQGpqKnJyciAI3GqViKir8dolVVtbCwDQarUu5RpN0wbiRqPR7esSExPbvG5xcTFMJhOKi4vx3HPPISIiAh999BEyMzNhNBrxn//5n6I+ABER+YbXwHD8ti+TydyWy+Xtm5kbHR2N7OxsDB06FL179wYATJw4ESaTCdnZ2XjiiSdahRQREUnH67e9Y8C6ZUuirq7O5fjt0mq1mDJlijMsHFJTU2GxWDx2dRERkTS8BoZj7KK0tNSlvKSkxOX47SosLMT27dvR2NjoUm4ymQAAERER7bouERF1Dq+BER8fD51Oh3379rmUHzhwAAkJCYiJiWnXG5eUlGDt2rU4cuSIS/nevXuh0+kQGxvbrusSEVHnEPUcxlNPPYVVq1YhLCwMqampOHjwIPLy8rBp0yYAgMFgQGlpKRITE0WPO6SmpmLEiBFYs2YNDAYD+vbtiw8++AAHDx7Ea6+91mrMhIiIpCUqMObPnw+LxYKcnBzs3LkTcXFx2LBhA2bNmgUAOHToEFatWoVt27ZhwoQJot5YrVYjOzsbmzdvRlZWFgwGA5KSkpCVlYXp06e3/xMREVGnkAk95KGHykoj7PaO/yibsl73uuPes7/+VYe/LxFRZ5HLZYiKuv1ZqFytloiIRGFgEBGRKAwMIiIShYFBRESiMDCIiEgUBgYREYnCwCAiIlEYGEREJAoDg4iIRGFgEBGRKAwMIiIShYFBRESiMDCIiEgUBgYREYnCwCAiIlEYGEREJAoDg4iIRGFgEBGRKAwMIiIShYFBRESiMDCIiEgUBgYREYnCwCAiIlEYGEREJAoDg4iIRGFgEBGRKAwMIiIShYFBRESiMDCIiEgUBgYREYnCwCAiIlEYGEREJAoDg4iIRGFgEBGRKAwMIiIShYFBRESiMDCIiEgUBgYREYnCwCAiIlEYGEREJAoDg4iIRGFgEBGRKAwMIiIShYFBRESiMDBEEgQBxoZG2O2C1FUhIpKEUuoKdHVmuwr7Py9FZY0JVpuAMUm9cPegKKmrRUTkc2xheGG0aXCjqgED+oUiUK2AodYsdZWIiCTBFoYXdqEpU5Pv6gNjQyOM9Y0S14iISBpsYXhhE+SQyQClQoaQYDVqGyxSV4mISBIMDC9sUECllEMmkyEkWAVLox3mRpvU1SIi8jkGhhd2QQ61UgEACAlWAQC7pYjILzEwvLAJCqhVTbfJERi19eyWIiL/Izowdu/ejdmzZ2PkyJHIyMjArl27RL/Jhg0b8Pjjj7cqt1qt2Lx5M6ZOnYpRo0bhkUcewdmzZ0Vf1xdsghwqZdNt0gapAQC1bGEQkR8SFRh5eXlYvnw5Jk2ahC1btmD8+PFYsWIF9u3b5/W1b731FnJyctwe+8Mf/oA333wTS5YswaZNm6BQKPD4449Dr9ff3qfoRHb80CWlUsoRqFagtoGBQUT+R9S02o0bNyIjIwOrV68GAEyePBnV1dV49dVXkZ6e7vY1N27cQGZmJvbu3YuQkJBWx8vKyvDOO+9gzZo1WLBgAQAgJSUFaWlpeOONN7B27dr2fqYOZRMUzhYG0NQtxTEMIvJHXlsYer0epaWlmDlzpkt5WloaioqKPLYGNm3ahPPnzyM3NxdDhw5tdfyzzz6DzWZDWlqas0ytViM1NRVHjhy53c/RaWyCHGqXwFBzDIOI/JLXwCgqKgIADBgwwKU8Pj4eAFBcXOz2dYsXL8aePXtw7733erxuWFgYIiMjW133+vXrMJlM3mvfyeyC0NQlpVI4y7RBKtSZrLDZ7RLWjIjI97wGRm1tLQBAq9W6lGs0GgCA0Wh0+7rExETI5Z4vbzQaW12z+XXr6uq8Va3Tmcw2ALJWXVIAUNdglahWRETS8BoYgtC0OqtMJnNb3lYoiLmu2PeTQoO5KRTUbgKD3VJE5G+8fts7BqxbtiQcLQB3A9piaLVat60IR5m71oevOQJD5dIlxam1ROSfvAaGY+yitLTUpbykpMTl+O0aOHAgbt68ierq6lbX1el0UKvV7bpuR6p308IIClBAqZAxMIjI73gNjPj4eOh0ulbPXBw4cAAJCQmIiYlp1xvfd999AID9+/c7yywWCw4fPuw8JjV3XVIymQzaIBWfxSAivyPqOYynnnoKq1atQlhYGFJTU3Hw4EHk5eVh06ZNAACDwYDS0lIkJiaK7kqKjY3FvHnz8PLLL6O+vh7x8fHIzc1FdXU1Fi9e3P5P1IGcXVJKhUu5NlgNI8cwiMjPiAqM+fPnw2KxICcnBzt37kRcXBw2bNiAWbNmAQAOHTqEVatWYdu2bZgwYYLoN3/ppZcQGhqKv/3tb6ivr8fw4cORm5vrnLIrNWcLQ+XaEAvTqHH9OyMarZxaS0T+QyZ4mq7UzVRWGjt8v+09x6/i/w4X4ZEZSVAqfgiN69/XIb+gDNPG6vBtYT6e/fWvOvR9iYg6k1wuQ1TU7U8s4mq1bag3WyGDAIXcdYpvn4ggyGUylFdK/6wIEZGvMDDa0GC2QQ5bq2dClAo5ekcEoryyXqKaERH5HgOjDQ1mKxQy9+MU/aI0qKo1wyoo3B4nIuppGBhtaAoM99ux9osKBgAYrcG+rBIRkWQYGG2oN1sh99DCiAoNhEoph9Gm8XGtiIikwcBoQ4PJCgXcB4ZcLkPfyGAYbWxhEJF/YGC0ob6NLimgqVuqUVDj25sNPqwVEZE0GBhtaGijSwoAYno1dUedvvy9r6pERCQZBoYHdrsAk8XmsUsKAEI1agTKTfj8fIUPa0ZEJA0GhgcmS9OyIG11SQFAuLIGxeW1uGHgMxlE1LMxMDxwLG3eVpcU0BQYMgCfn7/hg1oREUmHgeFBvcnRwmg7MFRyK4b0D8fx8zc87iJIRNQTMDA8cKxUq0DbXVIAMGFYNG4Y6lFyo7azq0VEJBkGhgcN5qag8NYlBQDJd/WBQi7DZ1+zW4qIei4GhgfOFoaXQW8A0ASqMCapF46dK3cOlhMR9TQMDA/qnV1S4jZJmjm+P+pMVhw9W96Z1SIikgwDw4MGkbOkHBJjwzAoNhQHvtR3+EZORERdAQPDgwazFUqFHHKZ+C//9PH98X21CScvfdeJNSMikgYDw4N6sxXBgaK2PHcak9QbfcKDsO+LUk6xJaIeh4HhQYPZiqCA2wsMuVyGtAn9UXS9Bv974BK7poioR7m9b0Q/Um+2IjhAAVhu73Wpo2PwfXUD8j4rRU2dBUvnDINaxV35iKj7YwvDA5PZhkD17eepTCbDQ6mJWDAtCScvfYe8z0s7oXZERL7HwPDAZLEhUN3+lsGMcXEYGBOKr4sNHVgrIiLpMDA8sDTaEHCHXUl39Y9AcXkNH+Yjoh6BYxgemBttUKsUqPNy3tmzZ7Ep63W3x2qtwbDZ++PKtWqMGBDV8ZUkIvIhBoYHZpEtDJO5EWNS5ro91mi142r+JVwsucnAIKJuj11SbgiC0BQY6ju7PSqlHMHyBlwsreqgmhERSYeB4YbVZocg4I7HMABAo6jH1fJa51IjRETdFQPDDXNj0/pRHfH8hFZRD7sg4HJZ9R1fi4hISgwMN8yWpiXNO6KFEaxogEIuY7cUEXV7DAw3zI0dFxhymYCBMaG4WMLAIKLujYHhRkcGBgCMGBiFqxW1+P5mQ4dcj4hICgwMNyzOwOiY2zNxWDQA4Ph5buFKRN0XA8MNRwtDfQdLgzTXKzwIg3VhOP5VBZc9J6Jui4HhhmOWVEd1SQHAxBF9UWGox9WK2g67JhGRLzEw3OjIWVIOyXf1gVIhw/GvKjrsmkREvsTAcKOjB70BQBOowqjEXvj8wg1YbeL2CSci6koYGG5YOiEwAOC+EX1RW9+I/V9wjwwi6n4YGG44WhiqDpol5TBqUC8k39UH/3e4CB8W6Dv02kREnY2r1brRtLS5HHKZrEOvK5fLsHTOMNjtAnbkX0aQWomUkf069D2IiDoLWxhumBvtHd4d5aBUyPHLucMxoF8o8j4v6ZT3ICLqDAwMN8yWO99try1KhRwThvZBeWU9n/4mom6DgeFGR2zP6s3dg5o2VDrHPb+JqJvgGIYbju1ZO4KnLVwFAVDJBmHvkTO4f0xsh7wXEVFnYmC40bQ9a8c0vtrawtXydQWu6A2w2uxQKtjYI6Kujd9Sbojdz/tOxfTSwA4FN1ciom6BgeGGudGOgA5aeLAt/aI0kEHAV0WVnf5eRER3ioHhhqUDxzDaolLKEayox7kiDnwTUdfHwHCjs6fVNheiqEPZd0Z8W1Xvk/cjImovBoYbvhrDAIBwZQ0Uchk+PnXNJ+9HRNReDIwWrDY7bHahw2ZJeaOSW3HP4N745Ey5cw0rIqKuiIHRQmetVNuWaWN1qDdb8Tm3cCWiLkx0YOzevRuzZ8/GyJEjkZGRgV27drV5fl1dHdauXYtJkyZhzJgxWLJkCa5evepyTkFBAYYMGdLqz7Jly9r1YTqCY7e9jtqeVYwkXRh0vbX46EQZt3Aloi5L1IN7eXl5WL58ORYtWoTJkycjPz8fK1asQGBgINLT092+5tlnn8W5c+fw/PPPQ6PRICsrC4sWLcKePXsQEhICACgsLERwcDByc3NdXhsaGnqHH6v9OmPzJG9kMhmmjY3FP/YV4nJZNQbHhfvsvYmIxBIVGBs3bkRGRgZWr14NAJg8eTKqq6vx6quvug2MgoICHD58GNnZ2ZgyZQoAIDk5GdOmTcOOHTuwdOlSAMDFixeRlJSE0aNHd9TnuWOdsT2rGPcO64t/HynCjvzL+O2isXzym4i6HK/fSnq9HqWlpZg5c6ZLeVpaGoqKiqDXt94I6NixY9BoNJg0aZKzLDIyEuPGjcORI0ecZRcuXMCQIUPupP4dTooWBgAEqBVYlDYEJTdqsec4lz0noq7Ha2AUFRUBAAYMGOBSHh8fDwAoLi52+5r4+HgoFK5fuv3793eeb7fbcfnyZVRUVGDevHkYMWIEUlNTkZOTI2k/vhSD3g5jh/TBvcOjsfvTq7haUePz9yciaovXwKitrQUAaLVal3KNRgMAMBqNrV5jNBpbne94jeP84uJimEwmFBcXY8mSJcjOzsb06dORmZmJ11577fY/SQdxtDDUPppW29LCGYMREqxC7t6LHAAnoi7F6xiG40tL1mK7Uke5XN76i7WtLzrH+dHR0cjOzsbQoUPRu3dvAMDEiRNhMpmQnZ2NJ554wm3odDapuqQcNIEqPDh5IN7Mu4hvrtUgURcmST2IiFryGhiOGU0tWxJ1dXUux5vTarUoKytrVV5XV+cMAa1W6xwQby41NRU7d+5EcXEx7r77bhEfoWM5p9X6KDDc7ZdhE2SQIxFb//kRRvUzYdHCR3xSFyKitngNDMfYRWlpqcsAdUlJicvxlq85fvw4BEFwaZmUlJQ4zy8sLMSJEyfw0EMPQaVSOc8xmUwAgIiIiPZ8njvm61lSnvbLaPyqAsXlSnxrKPRJPYiIvPHaUR8fHw+dTod9+/a5lB84cAAJCQmIiYlp9ZqUlBTU1NTg008/dZYZDAYUFBTgvvvuA9AUHmvXrnWZNQUAe/fuhU6nQ2ysNLvQOQe91dJOa03ShcFqE1Btbd2CIyKSgqjnMJ566imsWrUKYWFhSE1NxcGDB5GXl4dNmzYBaAqD0tJSJCYmQqvVYty4cRg/fjyee+45LF++HOHh4XjttdcQEhKCBQsWAGjqehoxYgTWrFkDg8GAvn374oMPPsDBgwfx2muvtRoz8RVzow1KhQwKN2MzvhQVFohwrRqGej7ER0Rdg6jAmD9/PiwWC3JycrBz507ExcVhw4YNmDVrFgDg0KFDWLVqFbZt24YJEyYAALKysrB+/XpkZmbCbrdj7Nix2Lx5M8LCmgZx1Wo1srOzsXnzZmRlZcFgMCApKQlZWVmYPn16J31c73y5Um1bZDIZknTh+PKiBSUVtYjvy5YGEUlL9J7eP/vZz/Czn/3M7bH58+dj/vz5LmVhYWFYt24d1q1b5/GakZGReOmll8RWwSfMPto8SYxBsaE4cbEc7x8rxtM/GSl1dYjIz3H9iRbMjfYu0cIAmmZq9VIbcOry93yQj4gkx8BowdJFuqQceqmqoAlUYtcnrZ+oJyLyJQZGC03bs3ad26KQ2ZFxbzzOflOJK9eqpa4OEfmxrvPN2EWYG20+3QtDjGn36BAarML7x9jKICLpMDBa6CqzpJoLUCvwo7E6fFVkQHllndTVISI/xcBooauNYTikjo6FUiFHfkHrJVeIiHyBgdFCV5ol1VyoRo17h0Xj2FflqDM1Sl0dIvJDDIwWumKXlMP0ZB0sjXYcOXNd6qoQkR9iYDRjtwtotNol2wvDm/7RIbirfzgOniiD1WaXujpE5Ge65jejRJx7YXSxWVLNpU/oj8oaMw6fZiuDiHxL9NIg/kDK7Vk9ablfhiAAGkUc/vnheZz6dDf6RIZyvwwi8gkGRjNS77bnjrv9MuJrTNjzaQnQewIqvz8uUc2IyN+wS6oZ063NkwK7cJcUAESFBmJQbCgultyExa7y/gIiog7AwGimrqFpuqomsOt/CY9J6gW5HCgxxaLBbJW6OkTkBxgYzdSZmr54tUFdPzCCA1WYMjoGJnsAtu76irOmiKjTMTCaMd56IE7TDQIDAHS9tYgNqMBXxQZs//CS1NUhoh6OgdHMD11S3WcuQKSqGtPu0eHwmeuoqjVLXR0i6sEYGM3UNVihUsq7zI57Yk0fp4MgAMfOlUtdFSLqwRgYzRhNjd2qdeEQHRGMIXHhOHq2HHZBkLo6RNRDMTCaqWto7BYD3u5MHtUP395swKXSm1JXhYh6KAZGM3UNjd1iSq07Y4f0QVCAAp+c5ZIhRNQ5GBjN1Jms3WaGVEsBKgUmDOuLgsLvUM/lz4moEzAwmumuYxgOU0fFoNFqx4fcZImIOkH3/XbsYIIgoK7B2u3GMFouThimjMEHR79B4ckDUMltiIoI5+KERNQhGBi3WKx2WG32btcl1XJxwkF1Frx3tBj2qGSMGd4Xp46+J2HtiKgnYZfULd3xoT13QjVqDIkLx+WyalQb+SAfEXUcBsYtxm608KA3IxOjoJTLcej0dZhsAVJXh4h6CAbGLY6FB7tbl5Q7gWolUu+Jgdliw5WGeOQX6KWuEhH1AAyMWxxdUt1t0NuTflEa/DglAVpFPbbnX8YXF25IXSUi6uYYGLfUmXrGGEZzgWol4gPLMKBfCP53fyEMNSapq0RE3RgD4xbnGEYPaWE4yGTAkjnD0Wiz4+97LnCtKSJqNwbGLXUmK5QKOdTKnndL+kYGY8G0JFwoqcL7R4ulrg4RdVM9p//lDjUtPKiETCaTuiqdYsqoGFwpq8b7x65CG6TC9OQ4qatERN0MA+OW7ryOVFuaPwkuCECoIhbb8y/jyCeHEaGq4ZPgRCQaA+OW7rxSbVtaPgk+ymbHRyeuocwQg0jd3fi+4piEtSOi7qTnddi3U3dfeFAshUKOaWNjkRgbhrPfVKLEpEOD2Sp1tYioG2Bg3FLX0Ngju6TcUSjkmDgiGuOH9UGtTYNNO88wNIjIKwbGLXWm7rdS7Z2QyWS4q38E+gdeR9G1GoYGEXnFwABgabSh0Wr3iy6plsKUtfjl3OEoulaDzB2nUFXLBQuJyD3/+4Z0oyetI3W7zp49CwCIC9BAXxGLla8fQnxgGYIUTcHBWVRE5MDAwA9PeWt74Cwpb5rPojLUmHDw5DWUWAZhxjgdeoUHcT8NInJilxSa7YXhhy2M5iJDAzHr3v4IUCuQf6KM3VNE5IKBgZ658GB7BQeqMGOcDkq5HB9+qUe9LVDqKhFRF8HAwA9jGP40S6otIcFqzBing0wGfNMQj5w9F3Dt+zrUm6wQuHghkd/ir9Rovj0rA8MhTBuAuSkDkH/kcxz/Wo6j58oBAIFqBWaOi0PGvfEIUCkkriUR+RIDA02D3kqFDGoVG1zNqVUK9Av4Dv/v5w/gkv4mausb8c31pgUMj54rx0/vT8SFEx/BcPOmx2twlhVRz8HAAFD6rRHRkcE9dqXaO3H27Fm89VauS9nAoCBcN0bjr+99DZgUmH1/GqLC3I91cJYVUc/h94Fhs9txpawak+7uK3VVuqSWixc62AUBV8qqcfyMCXuOlyBcq0ZCv1AMign1+9lmRD2V3/fBlFQYYW60YXBcuNRV6VbkMhkGx4XD9M37GHdXH6hVCpy+/D3+faQIh09fh/5bI+pMjeAYOVHP4fctjEv6pv73IQyM9rE3YmhCBIYmRMBY34hCfRUu66tRUlELAFAgCevfPom43loM7h+OkQOjEKDmYDlRd+T3gVFYWoXoyGCEaQOkrkq3pw1WYeyQPhiV2AuVNSZU1Zpx5VIh7PZIHPuqHB+dLINaJcfIgVFIvqsPRg3qxfAg6kb8OjDsdgGXyqox7q7eUlelR1Eq5IiOCEZ0RDA+3/0hYgNHIlEF1MmDUW0NwelLFhQUfgcZ7OgdbMZjP56IIfERkHPSAVGXJjowdu/eja1bt0Kv1yM2NhbLli3Dgw8+6PH8uro6/OlPf8KBAwdQX1+P5ORk/Pa3v0VCQoLzHKvViqysLLz77ru4efMmhg8fjpUrV2LkyJF39KHEKvvOiAazFUPiInzyfv7I3aC5XRDwbVUDrpbX4IregFf+eRrhWjVGDIjCsAERiOutRZ+IYKiUfj/ERtSliAqMvLw8LF++HIsWLcLkyZORn5+PFStWIDAwEOnp6W5f8+yzz+LcuXN4/vnnodFokJWVhUWLFmHPnj0ICQkBAPzhD3/Au+++i+XLlyMmJga5ubl4/PHH8d577yEuLq7jPqUHhbfGLzjg7VtymQx9I4PRNzIYKsMXmDRtHk5d+h4nL33nfEBQJgNiemkwKCYMCf1C0Ds8CL1CAxEUoIRaJYdaqYBczhYJkS+JCoyNGzciIyMDq1evBgBMnjwZ1dXVePXVV90GRkFBAQ4fPozs7GxMmTIFAJCcnIxp06Zhx44dWLp0KcrKyvDOO+9gzZo1WLBgAQAgJSUFaWlpeOONN7B27dqO+oweXdLfRK+wQI/PEFDn++rcGchlTVOpBioBkzwAZnsATHY1qg1GfPp9LY6ccd/SUCpkUCsVUKnkCFAqEKpVo3dYIPpEBKN/Hy36R4cgMjSAz9cQdRCvgaHX61FaWornnnvOpTwtLQ15eXnQ6/WtWgPHjh2DRqPBpEmTnGWRkZEYN24cjhw5gqVLl+Kzzz6DzWZDWlqa8xy1Wo3U1FQcOnToDj+WOCUVtWxdSMzTcx4O/3j997hrxD2wCCo02lWwC3LYIYcdMgiCHFWGagSERcIMGcprVCi5pkKjoATQFBJKuR1JcVGIidIgPESNkGA1gKbxK5tdcP4tCALsggC5XAalQn7rjwwqx38r5QgNViMqNAAhGjXHW8gveQ2MoqIiAMCAAQNcyuPj4wEAxcXFrQKjqKgI8fHxUChcZ8D0798feXl5znPCwsIQGRnZ6rrXr1+HyWRCYKD43/zb0z3xiweGoV9UcJuvDQ8PRXAba0xF9+nN4514PCwsHPfP9DxWtv3vG/HwQtdfZqw2O2qMFlTVmXG1uAiBAUpcuVaNBkvHbUGrgA1ymR0KhRxhoSFoyg8Z5Gja/lYma/qZlMkAyBzlTTHWVN70D7lM9kMZmp0va7pO09/44frOctmt8ls/u86/fvhZ9pRpMrQ+18N/un+h20PuD7rUwbWqLsdatgJbXU3W5j9dCB7/4frPthbSbHmozceJPBxsfQ2h+T88X6KN17V8q4ToENw9KKqt2rnV3u5cr4FRW9s0n16r1bqUazQaAIDRaGz1GqPR2Op8x2sc57d1DtA0aH47gRERoRF9rkNKVOv3b+n5Z59s8/jsyVt4vAsfB+7xcpyIxPI6DcWRwi1/A3CUy+WtL9FWcjvO93SOp/cjIiJpeQ0Mx4ymli2Juro6l+PNabVa5/GWr3G0Kto6x3GciIi6Dq+B4Ri7KC0tdSkvKSlxOd7yNXq9vlUroqSkxHn+wIEDcfPmTVRXV7c6R6fTQa1W38bHICKizuY1MOLj46HT6bBv3z6X8gMHDiAhIQExMTGtXpOSkoKamhp8+umnzjKDwYBaiBHjAAAIV0lEQVSCggLcd999AOD8e//+/c5zLBYLDh8+7DxGRERdh+LFF1980dtJISEh2Lp1K6qqqiCTyZCbm4t3330XL7zwApKSkmAwGFBYWAitVgu1Wo3Y2Fh88cUX2L59O8LDw3H9+nWsXr0agiDgj3/8IwIDAxEaGopr167h73//O4KCglBVVYWXXnoJer0emZmZCA/ndFcioq5EJojcpPmf//wncnJyUF5ejri4OCxdutS5NMi///1vrFq1Ctu2bcOECRMAANXV1Vi/fj3y8/Nht9sxduxYrFy5EgMHDnRe02Kx4E9/+hN2796N+vp6DB8+HM8//zxGjRrVCR+ViIjuhOjAICIi/8bV3YiISBQGBhERicLAaGb37t2YPXs2Ro4ciYyMDOzatUvqKnUJdrsdO3bswJw5czBmzBhMnz4d69atc3k259y5c3j00UcxZswYpKSkYOPGjWhsbJSw1l3Dr3/9a8yYMcOl7OjRo/jJT36CUaNG4Uc/+hFycnIkql3X8OWXX2LBggUYNWoUUlJS8Pvf/97lGS3erx/s2LEDGRkZGD16NObMmYP333/f5Xhn3ytRs6T8QV5eHn7zm99gzpw5ePLJJ2E2m7F582YkJSUhMTFR6upJKjs7G5mZmZg/fz6WLVuGhIQE/OMf/8DJkycxd+5clJSUYMGCBdDpdFixYgUSEhKwdetWGAwGTJ06VerqS+a9997D//zP/yAsLAyLFi0CAJw8eRK/+MUvMH78eDzzzDMICQnBX/7yFwQFBeGee/xvGZPTp0/jsccew4gRI7By5UoMHjwYb775Jq5cuYL09HTer2beeecdvPDCC5g3bx5+9atfwW6345VXXkFiYiKSkpJ8c68EEgRBEKZPny4888wzLmX/9V//JaSnp0tUo67BbrcL48aNE1588UWX8j179giDBw8Wzp8/L6xevVqYOnWqYDabncfffvttYejQoUJFRYWvq9wlVFRUCOPGjROmTJkiTJ8+3Vn+2GOPCQ899JDLuZmZmUJycrLL/fMXCxcuFBYuXCjY7XZn2VtvvSVMmzZNqK+v5/1q5uGHHxYeffRRl7JHHnlE+PnPfy4Igm9+ttglhR+WcJ85c6ZLeVpaGoqKiqDX6yWqmfTq6urw4x//GA888IBLuWN6dGlpKY4dO4b777/f5en89PR02Gw2HD161Kf17Sp+97vfYdKkSZg4caKzzGw2o6CgwO3PWU1NDU6ePOnrakrK8TDvggULXNaOW7hwIfLz8yGXy3m/mjGbzc7FWR3Cw8Nx8+ZNn/1sMTAgbgl3f6XVavG73/0OY8eOdSnPz88HAAwaNAjl5eWt7l1kZCS0Wq1f3rudO3fi66+/xpo1a1zK9Xo9Ghsb+XN2y6VLlyAIAsLCwvDMM89g9OjRGDt2LF544QWYTCberxYWLVqETz75BHl5eTAajdi3bx8OHTqEuXPn+uxeid7TuydrzxLu/uzMmTP429/+hunTpyM0NBSA+8Uimy9n7y+uXbuGdevWYd26da32euHPmSuDwQAAWLlyJWbMmIGtW7eisLAQmzdvhtlsxsMPPwyA98th9uzZ+Oyzz/DMM884y+bNm4fFixfj1KlTADr/XjEw0L4l3P3ViRMn8Mtf/hI6nQ4vv/wyLBYLAPfL0QuC4Ff3ThAErF69GlOnTnXZSbL5ccDz0v3+dK8AOGfR3XPPPXjhhRcAABMnToQgCNiwYQN++tOfAuD9cnjyySdx6tQprFq1CsOGDcOZM2fw+uuvQ6vVYtasWQA6/14xMNC+Jdz90d69e7Fy5UokJCTgjTfeQEREhPMeufsNpr6+3q/u3dtvv43CwkJ88MEHsFqbdvdzhITVavX4c+b4tz/dK+CH336nTJniUp6SkoL169fj3LlzAHi/gKbZdUePHsW6deswf/58AMD48eMRGhqK//7v/8Z//Md/AOj8e8XAgOsS7kOGDHGWt7WEu7/Jzc3Fhg0bMH78eGzZssX5A6jRaBAdHe28Vw6VlZUwGo1+de/279+PqqoqpKSktDo2fPhwvPjii1AoFK22CnD825/uFQAkJCQAgLOV6uBoeeh0Ot6vW65fvw4ArabHJicnAwAuXLjgk3vlX206D9qzhLs/2blzJ9avX4+MjAy88cYbrX5bmTRpEj7++GOX//H3798PhUKB8ePH+7q6klm7di3+9a9/ufy5//770bdvX/zrX/9Ceno6kpOTceDAAZe9Yvbv34+QkBCMGDFCwtr73qBBgxAbG4u9e/e6lH/88cdQKpUYM2YM79ctji/8L7/80qX89OnTAJpmLfriXvHBvVu8LeHuryorK7F48WJER0fjN7/5DSorK1FRUeH8o1arMWzYMOTk5KCgoABhYWE4dOgQXnnlFTz00EOYM2eO1B/BZyIiIhAdHe3y5+jRo/j222+xfPlyBAUFoW/fvvjrX/+Kb775BkFBQdi1axeys7Px9NNPO1d69hcymQy9e/dGbm4url69Cq1Wi7y8PGzZsgWPPvooZsyYwft1S58+fXDx4kVs374dgYGBMJvN2L9/P/785z9j4sSJWLJkiU/uFVerbaatJdz91a5du7BixQqPxzMzMzF37lwUFBQgMzMTFy5cQEREBB588EE8/fTTUKlUPqxt17Ny5UqcOHECH374obPsww8/xF/+8hcUFxcjOjoaCxcuxBNPPCFhLaWVn5+PLVu24MqVK4iKisLDDz+MZcuWOQdqeb+aWCwWZGVl4f3330dlZSViY2PxwAMPYOnSpc5noDr7XjEwiIhIFI5hEBGRKAwMIiIShYFBRESiMDCIiEgUBgYREYnCwCAiIlEYGEREJAoDg4iIRGFgEBGRKP8ff7+55laP2p0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(nums_trans_per_gene, hist_kws={'edgecolor':'black'}, bins = int(180/5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, 'number of transcripts per gene')"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkgAAAFSCAYAAAAJuyDPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3XlY1NXix/EPoKSpaWrWU25lDiUEkqKZGkWuld40l+uCFlqa5dIqdr1m2r1mi17T3K77VXEpcUkNM5dC0yfT1MqsW4lLWVcRr0uCwPn9wW8mvodFmTuCOe/X8+jDnPM9y5z5wnw4850hwBhjBAAAAI/Akp4AAADA5YaABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgKWUtw1PnDij7Gzjy7kAAAD4VGBggK69tlyR23kdkLKzDQEJAABckXiJDQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALKVKegK+tHDhPB06lOKz/k6eTJMkVaxYyWd9FqZGjVrq3r1XsYwFAAAKdkUFpEOHUrT/u38rqIxvAk3WuZyA9J//Zvqkv4sZCwAAlLwrKiBJUlCZSrq61v0+6etsykeS5LP+LmYsAABQ8rgGCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsl21A2rLlY23Z8nFJTwPFjMcdAHA5KFXSEyhIcvJmSVLTpveU8ExQnHjcAQCXg8t2BwkAAKCkEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkPCHERfX3fPvYuu8aXP59NdDcXHd1adPzzx1I0a8pLi47ho1arijfPLktxUX113Tp0/K02bgwH6Ki+uuwYOfzFPnTbv331+huLju+uCDVXnabNjwoeLiumvz5o/y1BXUbunSRYqL667ExCV52nz11R716dNDX3/95UWPVVibgwcP6Kmn+ujQoZQ8ddu3b1VcXHd99tk2R3la2gm99toonTyZVqT+CqorrI03YxXWxtu6gnjTn7f315s5+Lqdt2N5ozjH8iVv513YeVHSCEjAZcvk/G+y89QcPnxAknTgwA+O8h07cp7Ut23bmqfNmTOnJEmnTp3MU+dNu2XLFkuSlixJyNNm/vw5kqS5c2flqSuo3dq1KyVJq1Ytz9NmypSJMsZo8uQJFz1WYW2mT39Hv/32m6ZNyxsIZ8yY+v/HTHaUr1qVqO++26+VK5cVqb+C6gpr481YhbXxtq4g3vTn7f31Zg6+buftWN4ozrF8ydt5F3ZelDQCEv4Q8tt5uVCdN20un/56OOpy7yKNGPGSo869izR58tuO8ty7QQMH9nPU5d4N8qbd+++vcJTn3g3asOFDucOdZBw7OwW1W7p0kaM89y7SV1/t0dmzZyRJZ8+ecewIFTRWYW0OHjygn346Ikn66acjjt9ct2/fqqysTElSVlamZxcpLe2EkpM3yxij5OSPHb8lF9ZfQXWFtfFmrMLaeFtXEG/68/b+ejMHX7fzdixvFOdYvuTtvAs7Ly4HpUp6AgU5eTJNJ0+e1Nixoy+6zcGDKcrOCrqEs7p0sjPP6eDBlCLd3yvRwYMpqlixYklP4zJgnLdy7SK5d4/c3LtI7l0gt23btuqJJ56W9PsukFvu3SBv2rl3gdyWLElQmzbtJP2+o+M2d+4sRUffX2g79+6R26pVy9WhQxdJOTtBuU2ePEGTJv2z0LEKazN9+juOumnTJunVV9+Q9Pvukdv06ZMVFXWXVq1KVHZ2zmOSnZ2tlSuXKTY27oL9FVRXWBtvxiqsjbd1BfGmP2/vrzdz8HU7b8fyRnGO5Uvezruw8+JywA4SAB8zF7hdNO6doPxv5z9WYW3cv7Hmd9u9e2Tf/vTTLY6dpU8/3XJR/RVUV1gbb8YqrI23dQXxpj9v7683c/B1O2/H8kZxjuVL3s67sPPicnDZ7iBVrFhJFStW0tChf73oNmPHjta/Dx27hLO6dAJLlVHNGlWLdH+vRP6+g3ZlCJAzuAT8T71dfXU5R8C5+upyFxyrsDY33niT4wfxjTfe5Pk6KKiUIyQFBeX8iGzSpKk+/niTsrIyFRRUSk2aNL2o/gqqK6yNN2MV1sbbuoJ405+399ebOfi6nbdjeaM4x/Ilb+dd2HlxOWAHCbgsOUNFQMDv36rVq9d21NWufYskqWHDuxzld911t+frcuUqOOoqVPj9ZUxv2nXs2NVR3qVLN8/XPXs+6qjr3fv3rfaC2rVt295R3q7dw56vn3xyoKNuwIDBFxyrsDZPPPGUo65fv6c9X/ft299R98QTA/5/Ph0UGJjzmAQGBqp9+44X1V9BdYW18Waswtp4W1cQb/rz9v56Mwdft/N2LG8U51i+5O28CzsvLgcEJPwhzJq1sMDbBdV50+by6W+Bo27mzPmer0eN+rujbsSIVyVJAwYMcpS7ryOSpIkTpznqJkyY4vnam3YPPfQnR7n7+iNJiolpqd8DXoDn+qPC2nXu/GdHufv6I0kKDQ337ABdfXU51asXdsGxCmtTs2Ztx05OjRq1PHWNG9/t2TUKCiqlqKic8Fip0rVq1ixaAQEBatbsHlWsWOmi+iuorrA23oxVWBtv6wriTX/e3l9v5uDrdt6O5Y3iHMuXvJ13YefF5YCABFy2cp74c+8eubl3kdy7R27u3aDcu0Bu7t2g3LtH/0s7925Q7t0jN/fOTu7dowu1c+8i5d49cnvyyYEKCAhw7ARdaKzC2jzxxFMqW7Zsvr+xuneR3LtHbu3adVDduiH5/nZcWH8F1RXWxpuxCmvjbV1BvOnP2/vrzRx83c7bsbxRnGP5krfzLuy8KGkBxhivrqA8fvy056r1S8F9LYo31yBdXev+Cx98Ec6m5Lxl2Ff9XWisW7kGyavHHQCAggQGBqhKlfJFb3cJ5gIAAPCHRkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAAALAQkAAMBCQAIAALAQkAAAACwEJAAAAAsBCQAAwEJAAgAAsBCQAAAALAQkAAAACwEJAADAQkACAACwEJAAAAAsBCQAAABLqZKeQEGaNYsu6SmgBPC4AwAuB5dtQGra9J6SngJKAI87AOBywEtsAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGApVdIT8LWsc2k6m/KRz/qS5LP+LjxW1Us+DgAAuLArKiDVqFHLp/2dPJmzPBUrVvJpv/mr6vP5AwAA7wQYY4w3DY8fP63sbK+aAgAAFIvAwABVqVK+6O0uwVwAAAD+0AhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISAAAABYCEgAAgIWABAAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYSnnbMDAwwJfzAAAA8Dlv80qAMcb4eC4AAAB/aLzEBgAAYCEgAQAAWAhIAAAAFgISAACAhYAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGApsYD0/vvv68EHH1R4eLjatm2r5cuXl9RUSsS+ffsUGhqqo0ePOsqTk5P1yCOPKCIiQjExMZo1a1YJzfDSys7OVkJCgtq1a6fIyEi1aNFCY8aM0enTpz3H7N27V7GxsYqMjFSzZs00btw4nT9/vgRnfWkYYzRnzhy1bt1a4eHhat++vVatWuU4xl/Oi9yefvpptWzZ0lHmL+uQmZmp8PBwhYSEOP5FRkZ6jvGXtZCkzz77TN26dVNERISaNWum0aNH68yZM556f1iL7du35zkfcv9LTEyU5B9rIUkJCQlq27at6tevr3bt2mnlypWOep+sgykBa9asMSEhIeZvf/ub+fjjj82IESOMy+Uya9euLYnpFLvvv//eNG/e3LhcLvPzzz97yj///HMTGhpqnn/+ebN582Yzbtw4ExISYmbMmFGCs700pk2bZm6//Xbz5ptvmi1btpj58+ebRo0ambi4OGOMMQcOHDB33nmn6dOnj9m0aZOZOXOmCQsLM6+88koJz9z3pkyZYm6//XYzefJks3XrVvPaa68Zl8tlVq9ebYzxr/PCbfny5cblcpkWLVp4yvxpHb799lvjcrlMYmKi2bVrl+ff7t27jTH+tRa7du0yoaGhZuDAgWbLli1m4cKFpmHDhmbIkCHGGP9Zi1OnTjnOhV27dpmdO3eatm3bmujoaHP8+HG/WYtFixYZl8tlXnvtNbNlyxbz97//3bhcLrNmzRpjjO/OiRIJSC1atPCc3G6DBw82bdq0KYnpFJvz58+b+fPnm8jISNOoUaM8Aal3796mc+fOjjavv/66adiwoUlPTy/u6V4y2dnZJioqyowcOdJRvnr1auNyuczXX39tXnrpJRMdHe243wsWLDC33367OXr0aHFP+ZLJyMgwUVFRZtSoUY7ynj17mm7duhlj/Oe8cDt69KiJiooy99xzjyMg+dM6rFy50tx2223m7Nmz+db701r06NHD9OjRw2RnZ3vK5s+fb+6//35z9uxZv1oL25w5c8xtt91mvvjiC2OM/5wXXbt2NbGxsY6y7t27m549expjfLcOxf4S26FDh3Tw4EG1atXKUd66dWv98MMPOnToUHFPqdh8/vnnevPNNxUXF6fnn3/eUZeenq4dO3bkuy7//e9/tXPnzuKc6iV15swZtW/fXg899JCj/JZbbpEkHTx4UFu2bNF9992n4OBgT32bNm2UlZWl5OTkYp3vpRQUFKR//etfeuKJJxzlpUuXVnp6ul+dF27Dhw9X06ZN1aRJE0+Zv63Dvn37VLNmTZUtWzZPnT+tRWpqqnbs2KFu3bopIOD3v8jeo0cPrV+/XoGBgX6zFrZjx45pwoQJnpce/em8SE9PV7ly5RxllSpVUlpamk/XodgD0g8//CBJuvnmmx3ltWrVkiT9+OOPxT2lYlOnTh2tX79eTz/9tIKCghx1hw4d0vnz5/1iXcqXL6/hw4erQYMGjvL169dLylmnn3/+Oc9aVK5cWeXLl7+i1iIwMFAhISG6/vrrZYzRsWPHNH36dG3dulVdu3b1q/NCkpYuXaqvvvpKf/3rXx3l/rYO+/fvV3BwsPr06aPIyEhFRUVpxIgROn36tF+txbfffitjjCpWrKghQ4aofv36atCggV5++WWdO3fOr9bC9vbbbyswMFBDhgyR5F/fI7169dInn3yitWvX6vTp0/rggw+0adMm/elPf/LpOpTy6awvwqlTpyTlPEnm5k6DuS/SvdJUrVq1wDp/XhdJ2r17t6ZPn64WLVrommuukZR3LaSc9bhS12LdunUaNGiQJOnee+9V+/bttW/fPkn+cV4cOXJEY8aM0ZgxY1S5cmVHnb99f3zzzTc6ffq0OnfurP79++vLL7/UxIkT9eOPP+rZZ5+V5B9rkZqaKkmKj49Xy5YtNWXKFO3fv1//+Mc/lJ6erq5du0ryj7XILTU1VcuXL1dcXJzn56U/fY88+OCD2rZtmyccSlKHDh3Ut29f7dq1S5Jv1qHYA5IxRpIc26W5ywMD/fOTBwpaF7creV0+//xz9e/fX9WrV9err76qjIwMSfmvhTHmil2LevXqaf78+dq/f78mTJigJ554wvMD4Eo/L4wxeumllxQdHa3WrVvnWy9d+evgNn78eFWsWFEhISGSpKioKFWpUkUvvPCCtmzZIsk/1sL9rtU777xTL7/8siSpSZMmMsZo7Nix6tKliyT/WIvclixZouzsbPXq1ctT5k/fI08++aR27dqlYcOGqV69etq9e7cmT56s8uXL64EHHpDkm3Uo9oBUoUIFSXlTnPstm+56f1PQurhvX6nrsmbNGsXHx6t27dqaMWOGrr32Ws+5kF/SP3v27BW7FjVq1FCNGjUUFRWl8uXLa+jQoZ4felf6ebFgwQLt379fq1atUmZmpqTff+BnZmb63fdHo0aN8pTde++9jtv+sBbu3/rvueceR3mzZs302muvae/evZL8Yy1yS0pKUvPmzR07rf7yPbJz504lJydrzJgx6tixo6Sc75drrrlGI0aMUKdOnST5Zh2KPVK6Xxc8ePCgozwlJcVR729q1qypoKCgPOvivn0lrsvs2bP17LPPqn79+lqwYIGqVasmKeeH4vXXX+85J9yOHz+u06dPX1FrkZaWpuXLl+uXX35xlNerV0+SdPjwYb84L5KSknTixAk1a9ZMoaGhCg0N1fLly3Xw4EGFhoZqx44dfrEOUs55vnTp0jxvWDl37pwkqUqVKn6zFrVr15Ykz66ym3tnqXr16n6zFm6//PKLvv76a7Vt29ZR7i/PIT/99JOknF3F3Bo2bCgp5w0OvlqHYg9ItWrVUvXq1fXBBx84ytetW6fatWvrxhtvLO4pXRauuuoqNWzYUOvWrfP85izlPHFUqFBBYWFhJTg731u6dKlee+01tW3bVjNmzMiT6ps2baqNGzc6fjAmJSUpKCgo39+u/6iys7MVHx+vxYsXO8rdL6PccccdfnFevPLKK3r33Xcd/+677z7dcMMNevfdd9WmTRu/WAcp56WBESNGaP78+Y7yNWvWKCgoSHfffbffrEWdOnV00003ac2aNY7yjRs3qlSpUoqMjPSbtXDbvXu3JOV5k4u/PIe4A85nn33mKP/iiy8k5bwb2lfrEDRy5MiR//uUi6ZChQqaMmWKTpw4oYCAAM2ePVuJiYl6+eWXVbdu3eKeTonYt2+fPvroIz322GOei8luuOEGTZ06Vd9//73Kli2r5cuX65///KcGDhyoxo0bl/CMfef48ePq27evrr/+ej333HM6fvy4jh496vkXHBysevXqadasWdqxY4cqVqyoTZs26Y033lDnzp3Vrl27kr4LPlO2bFmlpqZq3rx5KlWqlDIyMrRixQpNmjRJHTt21COPPOIX58W1116r66+/3vEvOTlZv/76q55//nmVLVvWL9ZByjkn0tLStGDBAmVnZys7O1srVqzQ22+/re7du6tdu3Z+sxYBAQG67rrrNHv2bB04cEDly5fX2rVr9c477yg2NlYtW7b0m7VwW7t2rfbu3asXX3wxT50/rEW1atX0zTffaOHChSpTpozS09OVlJSkt956S02aNNHjjz/uu3Uo4ucz+UxCQoJp2bKlCQsLM23btjWJiYklNZUS8d577+X5oEhjjFm3bp156KGHTGhoqImJiTEzZ84soRleOomJicblchX4b/ny5cYYYz777DPTuXNnExYWZpo3b27eeustk5GRUcKz972MjAwzffp006pVKxMWFmZatGhhpk2bZrKysjzH+MN5YRs6dKjjgyKN8Z91cJ8TrVu3NmFhYeb+++/363Piww8/NA8//LAJCwsz0dHRZvLkyX67Fi+//LJp3rx5gfX+sBbp6enmrbfeMtHR0SYsLMy0bt3aTJw40fEhkL5YhwBjcu1BAQAAoOT+WC0AAMDlioAEAABgISABAABYCEgAAAAWAhIAAICFgAQAAGAhIAEAAFgISEARhYSEKD4+vqSnUaiMjAwNGzZMd955p+68805t2LChwGPtv/l1pYiPj1dISIhXbTMyMvL8fTwA/oWABFyBlixZomXLlqlFixYaNmxYgX9/6L333tODDz5YzLMrHl27dtXrr79e5HZHjhxRu3btPH8PD4B/KlXSEwDge/v375ckjRgxwvO3/vLz2WefKT09vbimVawiIyMVGRlZ5HaHDx/WgQMHfD8hAH8o7CABV6Dz589LUqHhCABQMAIS/hBiYmI0YsQIrVixQg8++KDuuOMOtWrVSgsWLMhzXGxsbL7tc5fHxMRo1KhRWrp0qVq3bq3w8HA98sgj2rNnj/7zn/9o8ODBioyMVPPmzTV+/HhlZ2fn6XPq1Klq3ry5IiIi1KtXL+3ZsyfPMRs3btSf//xnRUREKCoqSgMHDtSPP/7oOCYkJET/+Mc/1L9/f4WFhemBBx5QZmZmgWuxfv16/fnPf1Z4eLgaNmyo/v3765tvvnH0l5iY6Pk6v/WQpNjYWMdx7uuqYmJiNHz4cL300ku64447dM899yg1NVXGGCUkJKhTp06KjIzUHXfcoTZt2mj69OnK/ScdL/axOnnypOLj43XvvfcqLCxMLVq00FtvvZVnR+v777/X4MGD1bhxYzVo0ECxsbHasWOH43706dNH48ePV2RkpJo0aaL9+/fnuQYpPj5eLVu21K5du9SxY0eFh4erTZs2SkhI8ByzbNky9erVS5I0bNgwR/uEhAS1a9dOERERaty4sZ566il99913BT5O7v5CQkK0Z88ePfnkk6pfv76aNWumMWPG6Ny5c45j09PTNX78eMXExCgsLEz333+/JkyYoIyMjDz9JSUlKSYmRhEREZo4cWKB458+fVqvvPKKmjVrpvr166t///76/PPPFRISomXLlnmOy87O1qxZs9SmTRuFhYWpefPmevXVV3X69GnPMdu3b1dISIi2bNmiV155RU2aNFFERIR69+7tOP8utj/gcsdLbPjD+OSTT/TBBx+oZ8+eqlq1qhYvXqxRo0apevXqio6OLnJ/H330kdatW6fevXvLGKMpU6Zo4MCBqlChgurWrav4+HitW7dOU6dOVe3atdWhQwdP26SkJAUHB6tXr14qXbq05s2bp169emnp0qWqW7eupJwns5deeklNmjTRCy+8oJMnTyohIUFdunTRkiVLdPPNN3v6mzt3riIiIjR8+HCdO3dOpUrl/625YMECjRo1SmFhYXr22Wd1+vRpLVy4UN26ddPcuXMVHh6u119/XUuWLNGOHTv0+uuvq2rVqvn21b9/f2VnZ3uOq1mzpqdu9erVuvnmm/WXv/xFx44dU+XKlTV+/HhNnTpVHTp0UJcuXXTmzBktX75cb731lq677jrH+lzMYzVkyBB9/fXX6tWrl6pVq6Zdu3Zp+vTpSktL0+jRoyVJBw4cUJcuXVSqVCn17NlTlStX1qJFi/TYY49pwYIFCg8PlyTt3LlTKSkpeuGFF3T48GHdeuut+d7ntLQ09e3bV9HR0erYsaPWrVunkSNH6r///a/69eunqKgo9e/fX1OnTlXXrl3VoEEDSdLKlSs1cuRIPfzww4qNjVVqaqrmzp2r2NhYffjhh6pQoUKh59rgwYNVrVo1Pffcc9q3b5/mzJmjf//735o5c6YkKSsrS/369dPOnTvVpUsX1alTR19++aWmTp2qffv2acqUKQoICPD0N2zYMMXGxqpChQqqX79+vmNmZWWpb9++2rNnj7p3765atWpp9erVGjBgQJ5j//KXv2j58uXq0KGDHn30UX3//fdKSEjQzp07lZCQoKuuuspz7PDhw1WtWjUNGDBAJ0+e1IwZM/T4449r48aNnvO2KP0Bly0D/AHcd999JiQkxOzbt89T9uuvv5qQkBDz7LPPOo7r2bNnvu1zl7v7++abbzxlY8eONS6XywwZMsRTdubMGRMaGuoYw+Vymdtvv93R9sCBAyY0NNQ8/fTTxhhjTp06Ze68807zzDPPOObx66+/mqioKDNgwABHfw0aNDAnT54sdA1SU1NNRESE6dSpk0lPT/eUHzp0yFPuNnToUONyuQrtr6Dj7rvvPnPbbbeZlJQUT1lGRka+9+fUqVMmLCzM9OvXz9FgzNZGAAALSklEQVT+Qo/VsWPHjMvlMjNmzHD0Fx8fb3r37u25PXjwYBMeHm4OHDjgWIcGDRqYQYMGGWOM6dmzp3G5XGbbtm2F3jf37VdffdVTlpmZaXr06GHCw8NNWlqaMcaYbdu2GZfLZd577z3PcX379jUPPvigo/9NmzaZBx54wOzYscMU5L333jMul8s88sgjjsds3LhxxuVymY8//thxnPu226JFi4zL5TIffvih47ihQ4cWOKZbYmKicblcZsmSJZ6yjIwM06lTJ8f9c9/fhIQER/tPPvnEuFwuM2fOHMdxjzzyiMnMzPQcN23aNONyuUxycnKR+gMud7zEhj+Mm2++Wbfddpvn9nXXXaeqVavq2LFjXvVXs2ZNx0so7h2dli1besquvvpqValSRf/5z38cbZs3b+5oW6tWLTVv3lzJycnKysrSli1bdPr0abVo0UKpqamef0FBQbrrrruUnJzseBktIiJC11xzTaHz/fTTT/Xbb7/pscceU3BwsKe8evXqat++vfbs2aNff/3Vq7Ww1axZ07GjVLp0aW3dulWjRo1yHHfixAmVL19eZ8+edZRf6LGqUKGCrr76ai1cuFBJSUme9mPGjNGcOXMk5bxMs3nzZkVHR6tWrVqevq699lotXLhQw4cP95SVKVNGUVFRF3Xf+vXr5/k6KChIvXr10rlz57R169YC29xwww364YcfNGnSJB0+fFiSFB0drdWrV3t2mQoTFxfneMwee+wxSfJ8/MK6detUuXJlhYaGOs6X6OhoBQUFadOmTY7+mjVrdsEx169fr4oVK6pjx46estKlS3vGdlu3bp0CAgIUHR3tGLtevXq67rrr8ozdqlUrBQUFeW7ffvvtkuT5Hilqf8DlipfY8IdRuXLlPGXBwcH5Xh90MapUqeK47f6hb48TFBTkuMZGkm655ZY8/dWsWVMbNmxQamqqDh48KEl65plnChw/NTVV1apVy3fM/LifmPMbu06dOpKkn376ydPn/8JeGynnyXXTpk366KOP9OOPPyolJUUnT56UpDzrc6HHKjg4WKNGjdJf//pXDRo0SMHBwWrUqJFatWqlhx9+WFdddZXS0tJ09uxZRzhyc7lcjtuVKlVSYOCFf9+rVKlSnpcc3f0fOXKkwHZPPfWUvvjiC02cOFETJ07UrbfeqpiYGHXu3NkRJAvifnxyz6NSpUqeMQ8ePKjU1FQ1adIk3/Y///yz43Z+j48tJSVF1atXd4QZKe/5c/DgQRljdO+99+bbT7ly5Ry37cfWHfzcj21R+wMuVwQk/GFczBNgQbKysvKUFXSdT+5rPYrC/QQRFBTk+Xr06NGqXr16vsdXrFjR87X9JFZU7oBSunTp/6kfN3s+xhi98MILev/999WgQQNFRkaqa9euioqKUu/evfO0v5jHql27dmrevLnWr1+vzZs3a+vWrUpOTtbChQu1dOlSz2N2MX1d7Prltz65H7eC3HDDDVqxYoW2b9+ujz76SJ988ommT5+u2bNna9asWWrUqFGRx83KyvLct6ysLNWuXVsvv/xyvu3t3cWLWZPz58/nG0Zy72RJOfe/XLlymjRpUr792NcLXWjsovYHXK4ISLiiBAYGOt71I0mZmZk6ceLERf2mf7Hy221ISUlRhQoVdO211+qmm26SlPPb9t133+04bvv27crOzs7zRHUh7j5/+OEHx8tX7jIp54n8UtixY4fef/99DRgwQIMHD/aUZ2ZmKi0tTTVq1ChSf2fOnNG+fftUt25dderUSZ06dVJGRobeeOMNzZs3T8nJybrnnntUpkwZpaSk5Gk/c+ZMHTt2TEOHDi3SuMeOHdOZM2ccwcH9mUf57VS5uT9XqkmTJp5dns8//1y9e/fWv/71rwsGpEOHDjl2blJTU3Xq1CnVrl1bUs7LpF9++aXuuusuRwA5f/68PvzwQ68e1xo1amjv3r0yxjhCv72eN910k5KTkxUWFpYniCUlJalSpUpFGtfX/QElhWuQcEWpWrWqfvzxR8dbqDds2ODzD0P85JNPHH+K4ttvv1VycrJiYmIUEBCgu+++W1dddZVmzJjh+UwiSfrll180YMAAvfnmm0XeqXL3OXv2bEcIPHr0qFatWqXw8PCLeuklN/eT8YVepkxLS5OkPO8OW7JkiX777bdCP5YgP99995169Oihd99911MWHBysevXqScrZzSlVqpSaNm2qzZs3O15iOnnypGbOnOl5GbMojDGOjxvIzMzU3LlzVaFCBU/wce8k5V6TwYMH68UXX3TsRNarV0+lS5e+qN2c+fPnO16GdL97zX29W0xMjNLS0hwfOSBJixYt0jPPPKNPP/20qHdVLVu21IkTJ7R27VpPWXZ2thYtWuQ4LiYmRpI0ZcoUR/mGDRs0aNAgrVq1qkjj+ro/oKSwg4QrykMPPaTRo0erb9++at++vVJSUrRkyRLP7ouvBAcHq3v37oqNjdVvv/2mOXPm6JprrtGQIUMk5ewcPfvssxozZoy6du2q9u3bKzMzUwsXLlR6enqRdz6knIuT3X1269ZN7dq105kzZ5SQkKDs7GzHRcsXy309ydtvv63GjRsXeA1MZGSkypcvrzFjxuinn37SNddco+3bt2vNmjW66qqrdObMmSKNGxERoYYNG2r8+PH6+eefFRISop9//lnz58/XLbfc4pnHc889p86dO6tz587q0aOHypcvryVLlujs2bOetS6qyZMn68iRI6pbt67Wrl2rXbt26W9/+5vKli0rKWedpZy39htj1KFDB/Xp00fDhw/Xo48+qjZt2sgYoxUrVig9PV3du3e/4Jjbt2/X448/rvvuu0+7d+/WihUr9PDDD3su8O7cubMSExM1evRoffXVVwoPD9e3336rxYsXKzQ01HGh9cXq0KGDFi1apBdffFG7du1S7dq1lZSUpN27d0v6/aXk6Oho3X///Zo1a5YOHz6su+++W0eOHNGCBQt04403qk+fPkUa19f9ASWFgIQrSvfu3ZWWlqZ3331Xo0eP1m233aZJkyZp1qxZed5p9b/o2rWrAgICNHXqVKWnp6tx48aKj4/XjTfe6Dnm0Ucf1fXXX6/Zs2dr/PjxKlOmjEJDQ/XGG29c1Duf8vPoo4+qWrVqmjVrlsaNG6eyZcuqUaNGevrpp736w6zdunXTtm3bNGPGDO3du7fAgFS1alVNnz5db775piZPnqzg4GDdfPPNGjdunPbs2aN58+bp2LFjBX7mki0gIEDvvPOOJk2apI0bN2rx4sWqWLGiWrVqpcGDB3tefqxTp44WL16scePGacaMGQoMDFR4eLjGjh3r+bypopo5c6ZGjhypxMRE3XrrrZo0aZLjnYt16tRRbGysli1bpr1796px48bq3Lmz5/Ouxo0bp+zsbIWFhemf//ynGjdufMEx//73vysxMVFjx47Vddddp+eee059+/b11AcHB2vOnDl65513lJSUpJUrV6patWrq1q2bnnrqKU94K4rSpUtrxowZeuONN7Ry5Uqlp6eradOmGjlypOLj4z1rHBAQoAkTJmjGjBlavny5Nm7cqMqVK3sei4t9TN183R9QUgKM/fYTALgCxcfHKzEx0XM9UXFYtmyZhg0bpnnz5l1UkPKltLQ0lStXLs8F4klJSRo0aJDmzJlTYCAGwDVIAHBFmjdvnurXr6+jR486ylevXq1SpUp5rvcCkD9eYgOAK1Dbtm01ffp0xcXFqUuXLipTpoy2bNmidevW6cknn3R8zASAvAhIAHAFqlu3rhYsWKBJkyZp2rRp+u2331S7dm2NHj1aXbp0KenpAZc9rkECAACwcA0SAACAhYAEAABgISABAABYCEgAAAAWAhIAAIDl/wCwMygvu5XsSAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, axes = plt.subplots(figsize=(10,5))\n",
    "sns.boxplot(nums_trans_per_gene, width=0.5, ax = axes)\n",
    "axes.set_xlabel('number of transcripts per gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "77"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(nums_trans_per_gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Protein coding gentype on gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique_unsorted_elements(array):\n",
    "    array = np.copy(array)\n",
    "    unique_elements, indexes = np.unique(array, return_index=True)\n",
    "    #prepared for sorting indexes\n",
    "    merged_table = pd.DataFrame(unique_elements, index = indexes)\n",
    "    merged_table.sort_index(inplace= True)\n",
    "    unique_unsorted_elements = np.array(merged_table[0])\n",
    "    return unique_unsorted_elements\n",
    "\n",
    "def compare_gencode_chr_portion(capture_list, gencode_table, chr_name):\n",
    "    #return gencode data with dropping unsatisfied data points\n",
    "    capture_chr_list = capture_list[capture_list[0]==chr_name]\n",
    "    #print(capture_chr_list)\n",
    "    capture_chr_list.sort_values(capture_chr_list.columns[1], axis = 0, inplace= True)\n",
    "    capture_chr_list.reset_index(drop= True, inplace = True)\n",
    "    gencode_chr_portion = gencode_table[gencode_table.seqname==chr_name]\n",
    "    gencode_range_array = np.array([gencode_chr_portion.start,\\\n",
    "                                   gencode_chr_portion.end]).T\n",
    "    #print(gencode_range_array)\n",
    "    ovl_indexes, percents_ovl = search_overlap_partners(np.array(capture_chr_list[[1,2]]),gencode_range_array)\n",
    "    #reset_ovl_indexes = np.array(gencode_chr_portion.index)[ovl_indexes]\n",
    "    #print(ovl_indexes)\n",
    "    #print(len(ovl_indexes)-len(gencode_table))\n",
    "    non_ovl_indexes = np.delete(np.array(gencode_chr_portion.index), ovl_indexes, None) #delete portion of array that has the index\n",
    "    #print(non_ovl_indexes)\n",
    "    gencode_table.drop(index = non_ovl_indexes, inplace= True)\n",
    "    return gencode_table\n",
    "\n",
    "def compare_gencode_chr_portion_wo_capture(ref_list, gencode_table, chr_name):\n",
    "    #return gencode data with dropping unsatisfied data points\n",
    "    ref_chr_list = ref_list[ref_list.seqname==chr_name]\n",
    "    #print(capture_chr_list)\n",
    "    ref_chr_list.sort_values(by = ['start'], axis = 0, inplace= True)\n",
    "    ref_chr_list.reset_index(drop= True, inplace = True)\n",
    "    gencode_chr_portion = gencode_table[gencode_table.seqname==chr_name]\n",
    "    gencode_range_array = np.array([gencode_chr_portion.start,\\\n",
    "                                   gencode_chr_portion.end]).T\n",
    "    #print(gencode_range_array)\n",
    "    ovl_indexes, percents_ovl = search_overlap_partners(np.array([ref_chr_list.start, ref_chr_list.end]).T,\\\n",
    "                                                        gencode_range_array)\n",
    "    #reset_ovl_indexes = np.array(gencode_chr_portion.index)[ovl_indexes]\n",
    "    #print(ovl_indexes)\n",
    "    #print(len(ovl_indexes)-len(gencode_table))\n",
    "    non_ovl_indexes = np.delete(np.array(gencode_chr_portion.index), ovl_indexes, None) #delete portion of array that has the index\n",
    "    #print(non_ovl_indexes)\n",
    "    gencode_table.drop(index = non_ovl_indexes, inplace= True)\n",
    "    return gencode_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>70006</td>\n",
       "      <td>70008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000003223.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>182393</td>\n",
       "      <td>182708</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279928.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.3-201</td>\n",
       "      <td>basic,appris_principal_1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485457.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>184156</td>\n",
       "      <td>184158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279928.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.3-201</td>\n",
       "      <td>basic,appris_principal_1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485457.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>184923</td>\n",
       "      <td>185219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-201</td>\n",
       "      <td>basic,appris_alternative_2</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485222.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>184925</td>\n",
       "      <td>185219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-202</td>\n",
       "      <td>basic,appris_principal_5</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485442.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname   source feature   start     end  score strand frame  \\\n",
       "59     chr1   HAVANA     UTR   70006   70008    NaN      +     0   \n",
       "138    chr1  ENSEMBL     UTR  182393  182708    NaN      +     0   \n",
       "139    chr1  ENSEMBL     UTR  184156  184158    NaN      +     0   \n",
       "164    chr1  ENSEMBL     UTR  184923  185219    NaN      -     0   \n",
       "190    chr1  ENSEMBL     UTR  184925  185219    NaN      -     0   \n",
       "\n",
       "               gene_id       gene_type     ...      transcript_status  \\\n",
       "59   ENSG00000186092.4  protein_coding     ...                  KNOWN   \n",
       "138  ENSG00000279928.1  protein_coding     ...                  KNOWN   \n",
       "139  ENSG00000279928.1  protein_coding     ...                  KNOWN   \n",
       "164  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "190  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "\n",
       "    transcript_name                            tag transcript_support_level  \\\n",
       "59        OR4F5-001  basic,appris_principal_1,CCDS                       NA   \n",
       "138  FO538757.3-201       basic,appris_principal_1                        1   \n",
       "139  FO538757.3-201       basic,appris_principal_1                        1   \n",
       "164  FO538757.2-201     basic,appris_alternative_2                        5   \n",
       "190  FO538757.2-202       basic,appris_principal_5                        5   \n",
       "\n",
       "        havana_transcript exon_number exon_id ont         protein_id  \\\n",
       "59   OTTHUMT00000003223.1                          ENSP00000334393.3   \n",
       "138                                                ENSP00000485457.1   \n",
       "139                                                ENSP00000485457.1   \n",
       "164                                                ENSP00000485222.1   \n",
       "190                                                ENSP00000485442.1   \n",
       "\n",
       "          ccdsid  \n",
       "59   CCDS30547.1  \n",
       "138               \n",
       "139               \n",
       "164               \n",
       "190               \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check if any UTRs in CAPTURE:\n",
    "UTR_regions = gencode_data[(gencode_data.feature=='UTR') & (gencode_data.gene_type=='protein_coding')]\n",
    "UTR_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ordered by gencode:\n",
    "chr_names = unique_unsorted_elements(np.array(UTR_regions.seqname))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ndnguyen20/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  del sys.path[0]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(216034, 26)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR_regions_in_CAPTURE = pd.DataFrame.copy(UTR_regions)\n",
    "#set(UTR_regions_in_CAPTURE.seqname)\n",
    "for name in chr_names:\n",
    "    UTR_regions_in_CAPTURE = compare_gencode_chr_portion(capture_data, UTR_regions_in_CAPTURE, name)\n",
    "UTR_regions_in_CAPTURE.shape\n",
    "#no UTR regions found in capture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7828112996970708"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR_regions_in_CAPTURE.shape[0]/UTR_regions.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>69091</td>\n",
       "      <td>69093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000003223.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002319515.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>182709</td>\n",
       "      <td>182711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279928.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.3-201</td>\n",
       "      <td>basic,appris_principal_1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003759020.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485457.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>195409</td>\n",
       "      <td>195411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-201</td>\n",
       "      <td>basic,appris_alternative_2</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003758618.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485222.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>195409</td>\n",
       "      <td>195411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-202</td>\n",
       "      <td>basic,appris_principal_5</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003755778.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485442.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>200084</td>\n",
       "      <td>200086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-203</td>\n",
       "      <td>basic,appris_alternative_2</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003757911.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485351.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname   source      feature   start     end  score strand frame  \\\n",
       "57     chr1   HAVANA  start_codon   69091   69093    NaN      +     0   \n",
       "132    chr1  ENSEMBL  start_codon  182709  182711    NaN      +     0   \n",
       "144    chr1  ENSEMBL  start_codon  195409  195411    NaN      -     0   \n",
       "168    chr1  ENSEMBL  start_codon  195409  195411    NaN      -     0   \n",
       "194    chr1  ENSEMBL  start_codon  200084  200086    NaN      -     0   \n",
       "\n",
       "               gene_id       gene_type     ...      transcript_status  \\\n",
       "57   ENSG00000186092.4  protein_coding     ...                  KNOWN   \n",
       "132  ENSG00000279928.1  protein_coding     ...                  KNOWN   \n",
       "144  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "168  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "194  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "\n",
       "    transcript_name                            tag transcript_support_level  \\\n",
       "57        OR4F5-001  basic,appris_principal_1,CCDS                       NA   \n",
       "132  FO538757.3-201       basic,appris_principal_1                        1   \n",
       "144  FO538757.2-201     basic,appris_alternative_2                        5   \n",
       "168  FO538757.2-202       basic,appris_principal_5                        5   \n",
       "194  FO538757.2-203     basic,appris_alternative_2                        5   \n",
       "\n",
       "        havana_transcript exon_number            exon_id ont  \\\n",
       "57   OTTHUMT00000003223.1           1  ENSE00002319515.1       \n",
       "132                                 1  ENSE00003759020.1       \n",
       "144                                 1  ENSE00003758618.1       \n",
       "168                                 1  ENSE00003755778.1       \n",
       "194                                 1  ENSE00003757911.1       \n",
       "\n",
       "            protein_id       ccdsid  \n",
       "57   ENSP00000334393.3  CCDS30547.1  \n",
       "132  ENSP00000485457.1               \n",
       "144  ENSP00000485222.1               \n",
       "168  ENSP00000485442.1               \n",
       "194  ENSP00000485351.1               \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check if start codons in CAPTURE:\n",
    "start_codons= gencode_data[(gencode_data.feature=='start_codon') & (gencode_data.gene_type=='protein_coding')]\n",
    "start_codons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ndnguyen20/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  del sys.path[0]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.985237687889527"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_codons_in_CAPTURE = pd.DataFrame.copy(start_codons)\n",
    "for name in chr_names:\n",
    "    start_codons_in_CAPTURE = compare_gencode_chr_portion(capture_data,start_codons_in_CAPTURE, name)\n",
    "start_codons_in_CAPTURE.shape[0]/start_codons.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>70006</td>\n",
       "      <td>70008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000003223.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002319515.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>184156</td>\n",
       "      <td>184158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279928.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.3-201</td>\n",
       "      <td>basic,appris_principal_1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003760199.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485457.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>185217</td>\n",
       "      <td>185219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-201</td>\n",
       "      <td>basic,appris_alternative_2</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>ENSE00003758954.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485222.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>185217</td>\n",
       "      <td>185219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-202</td>\n",
       "      <td>basic,appris_principal_5</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>ENSE00003755076.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485442.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>184927</td>\n",
       "      <td>184929</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>FO538757.2-203</td>\n",
       "      <td>basic,appris_alternative_2</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>ENSE00003759031.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000485351.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname   source     feature   start     end  score strand frame  \\\n",
       "58     chr1   HAVANA  stop_codon   70006   70008    NaN      +     0   \n",
       "137    chr1  ENSEMBL  stop_codon  184156  184158    NaN      +     0   \n",
       "163    chr1  ENSEMBL  stop_codon  185217  185219    NaN      -     0   \n",
       "189    chr1  ENSEMBL  stop_codon  185217  185219    NaN      -     0   \n",
       "223    chr1  ENSEMBL  stop_codon  184927  184929    NaN      -     0   \n",
       "\n",
       "               gene_id       gene_type     ...      transcript_status  \\\n",
       "58   ENSG00000186092.4  protein_coding     ...                  KNOWN   \n",
       "137  ENSG00000279928.1  protein_coding     ...                  KNOWN   \n",
       "163  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "189  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "223  ENSG00000279457.2  protein_coding     ...                  KNOWN   \n",
       "\n",
       "    transcript_name                            tag transcript_support_level  \\\n",
       "58        OR4F5-001  basic,appris_principal_1,CCDS                       NA   \n",
       "137  FO538757.3-201       basic,appris_principal_1                        1   \n",
       "163  FO538757.2-201     basic,appris_alternative_2                        5   \n",
       "189  FO538757.2-202       basic,appris_principal_5                        5   \n",
       "223  FO538757.2-203     basic,appris_alternative_2                        5   \n",
       "\n",
       "        havana_transcript exon_number            exon_id ont  \\\n",
       "58   OTTHUMT00000003223.1           1  ENSE00002319515.1       \n",
       "137                                 3  ENSE00003760199.1       \n",
       "163                                10  ENSE00003758954.1       \n",
       "189                                11  ENSE00003755076.1       \n",
       "223                                15  ENSE00003759031.1       \n",
       "\n",
       "            protein_id       ccdsid  \n",
       "58   ENSP00000334393.3  CCDS30547.1  \n",
       "137  ENSP00000485457.1               \n",
       "163  ENSP00000485222.1               \n",
       "189  ENSP00000485442.1               \n",
       "223  ENSP00000485351.1               \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stop_codons= gencode_data[(gencode_data.feature=='stop_codon') & (gencode_data.gene_type=='protein_coding')]\n",
    "stop_codons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ndnguyen20/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  del sys.path[0]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.9419801205712975"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stop_codons_in_CAPTURE = pd.DataFrame.copy(stop_codons)\n",
    "for name in chr_names:\n",
    "    stop_codons_in_CAPTURE = compare_gencode_chr_portion(capture_data, stop_codons_in_CAPTURE, name)\n",
    "stop_codons_in_CAPTURE.shape[0]/stop_codons.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quick search how many UTRs are stop codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "stop_codons.reset_index(drop= True, inplace= True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ndnguyen20/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.32062310669198324"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR_regions_in_stop_codons = pd.DataFrame.copy(UTR_regions)\n",
    "for name in chr_names:\n",
    "    UTR_regions_in_stop_codons = compare_gencode_chr_portion_wo_capture(stop_codons, UTR_regions_in_stop_codons, name)\n",
    "UTR_regions_in_stop_codons.shape[0]/UTR_regions.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Making sure CAPTURE exons do not overlap each other:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ndnguyen20/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "[]\n",
      "chr2\n",
      "[]\n",
      "chr3\n",
      "[]\n",
      "chr4\n",
      "[]\n",
      "chr5\n",
      "[]\n",
      "chr6\n",
      "[]\n",
      "chr7\n",
      "[]\n",
      "chr8\n",
      "[]\n",
      "chr9\n",
      "[]\n",
      "chr10\n",
      "[]\n",
      "chr11\n",
      "[]\n",
      "chr12\n",
      "[]\n",
      "chr13\n",
      "[]\n",
      "chr14\n",
      "[]\n",
      "chr15\n",
      "[]\n",
      "chr16\n",
      "[]\n",
      "chr17\n",
      "[]\n",
      "chr18\n",
      "[]\n",
      "chr19\n",
      "[]\n",
      "chr20\n",
      "[]\n",
      "chr21\n",
      "[]\n",
      "chr22\n",
      "[]\n",
      "chrX\n",
      "[]\n",
      "chrY\n",
      "[]\n",
      "chrM\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for name in chr_names:\n",
    "    capture_check = detect_collapse_hg38_exons(CAPTURE_dict_chr[name])\n",
    "    capture_check.sort_values(1, inplace= True)\n",
    "    capture_check.reset_index(drop= True, inplace= True)\n",
    "\n",
    "#--> we have a hundred of those having alternative haplotypes that collapse into a single exon\n",
    "# no end-to-end touching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000003223.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000003223.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002319515.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69091</td>\n",
       "      <td>70005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000003223.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002319515.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>69091</td>\n",
       "      <td>69093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000003223.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002319515.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000334393.3</td>\n",
       "      <td>CCDS30547.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  source      feature  start    end  score strand frame  \\\n",
       "53    chr1  HAVANA         gene  69091  70008    NaN      +     0   \n",
       "54    chr1  HAVANA   transcript  69091  70008    NaN      +     0   \n",
       "55    chr1  HAVANA         exon  69091  70008    NaN      +     0   \n",
       "56    chr1  HAVANA          CDS  69091  70005    NaN      +     0   \n",
       "57    chr1  HAVANA  start_codon  69091  69093    NaN      +     0   \n",
       "\n",
       "              gene_id       gene_type     ...      transcript_status  \\\n",
       "53  ENSG00000186092.4  protein_coding     ...                          \n",
       "54  ENSG00000186092.4  protein_coding     ...                  KNOWN   \n",
       "55  ENSG00000186092.4  protein_coding     ...                  KNOWN   \n",
       "56  ENSG00000186092.4  protein_coding     ...                  KNOWN   \n",
       "57  ENSG00000186092.4  protein_coding     ...                  KNOWN   \n",
       "\n",
       "   transcript_name                            tag transcript_support_level  \\\n",
       "53                                                                           \n",
       "54       OR4F5-001  basic,appris_principal_1,CCDS                       NA   \n",
       "55       OR4F5-001  basic,appris_principal_1,CCDS                       NA   \n",
       "56       OR4F5-001  basic,appris_principal_1,CCDS                       NA   \n",
       "57       OR4F5-001  basic,appris_principal_1,CCDS                       NA   \n",
       "\n",
       "       havana_transcript exon_number            exon_id ont  \\\n",
       "53                                                            \n",
       "54  OTTHUMT00000003223.1                                      \n",
       "55  OTTHUMT00000003223.1           1  ENSE00002319515.1       \n",
       "56  OTTHUMT00000003223.1           1  ENSE00002319515.1       \n",
       "57  OTTHUMT00000003223.1           1  ENSE00002319515.1       \n",
       "\n",
       "           protein_id       ccdsid  \n",
       "53                                  \n",
       "54  ENSP00000334393.3  CCDS30547.1  \n",
       "55  ENSP00000334393.3  CCDS30547.1  \n",
       "56  ENSP00000334393.3  CCDS30547.1  \n",
       "57  ENSP00000334393.3  CCDS30547.1  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_protein_coding_gene = gencode_data[(gencode_data.gene_type=='protein_coding')]\n",
    "gencode_protein_coding_gene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "make sure GENCODE exons do overlap each other per gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Report fractions of genes found in CAPTURE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Copy and modify from GENCODE_CAPTURE_matching.ipynb\n",
    "def percent_difference(reference, range_ovl):\n",
    "    reference = np.copy(reference)\n",
    "    range_ovl = np.copy(range_ovl)\n",
    "    rd = (min(reference[1], range_ovl[1])-max(reference[0], range_ovl[0])+1)/(reference[1]-reference[0]+1)\n",
    "    return rd\n",
    "def range_difference(range1, range2):\n",
    "    range1 = np.copy(range1)\n",
    "    range2 = np.copy(range2)\n",
    "    rd = min(range1[1], range2[1])-max(range1[0], range2[0])+1\n",
    "    return rd\n",
    "def search_overlap_partners(chr_ref, chr_search, return_indexes = False):\n",
    "    \"\"\"reform the search function\"\"\"\n",
    "    chr_ref = np.copy(chr_ref)\n",
    "    chr_search = np.copy(chr_search)\n",
    "    search_indexes = []\n",
    "    ref_indexes = []\n",
    "    p_refs=[]\n",
    "    ranges_ovl = np.array([0])\n",
    "\n",
    "    for i in range(chr_search.shape[0]):\n",
    "        \n",
    "        p_search = chr_search[i]\n",
    "        closest_ref_val = np.argwhere(chr_ref[:,0]-p_search[0]<0)\n",
    "        if closest_ref_val.size > 0:\n",
    "            ref_i = np.max(closest_ref_val) #lower start of ref closest to start of search \n",
    "            p_ref = chr_ref[ref_i]\n",
    "        else:\n",
    "\n",
    "            ref_i = 0 #just to make sure\n",
    "            p_ref = chr_ref[ref_i]\n",
    "            #not to let the ref pass over:\n",
    "        while ((max(p_ref[0], p_search[0])-min(p_ref[1], p_search[1]))>0) and (p_ref[0]<p_search[1]) and (ref_i < len(chr_ref)):\n",
    "\n",
    "\n",
    "\n",
    "            ref_i += 1\n",
    "            if ref_i <len(chr_ref):\n",
    "                p_ref = chr_ref[ref_i]\n",
    "\n",
    "        #Just to make sure a search is matched with all ref\n",
    "        while (p_search[1]>p_ref[0]) and (ref_i <len(chr_ref)):\n",
    "            if ((min(p_ref[1], p_search[1])-max(p_ref[0], p_search[0]))>0):\n",
    "\n",
    "                if i not in search_indexes:\n",
    "                    #print(p_search, p_ref)\n",
    "                    ranges_ovl= np.append(ranges_ovl, \\\n",
    "                                          range_difference(p_search, p_ref)) #percent of search being overlapped\n",
    "                search_indexes.append(i)\n",
    "                ref_indexes.append(ref_i)\n",
    "                p_refs.append(p_ref.tolist())\n",
    "            ref_i+=1\n",
    "            if ref_i <len(chr_ref):\n",
    "                p_ref= chr_ref[ref_i]\n",
    "\n",
    "            \n",
    "    if len(p_refs)>0: \n",
    "        unique_p_refs = np.unique(p_refs, axis = 0)\n",
    "\n",
    "        num_refs_ovl = unique_p_refs.shape[0]\n",
    "        if unique_p_refs.shape[0] > 1:\n",
    "            ranges_ref_ovl = unique_p_refs[:,1]- unique_p_refs[:,0]+1\n",
    "        else:\n",
    "\n",
    "            ranges_ref_ovl = np.array(unique_p_refs[0][1]-unique_p_refs[0][0]+1)\n",
    "    else:\n",
    "        num_refs_ovl = 0\n",
    "        ranges_ref_ovl = np.array([], dtype=int)\n",
    "\n",
    "    if return_indexes:\n",
    "        return ref_indexes, search_indexes\n",
    "    else:\n",
    "        return ranges_ovl, num_refs_ovl, ranges_ref_ovl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_collection = unique_unsorted_elements(gencode_protein_coding_gene.gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def temporary_table(a =0):\n",
    "    return pd.DataFrame(columns=[['gene_name','gene_id','#TX','#CAP_exons','#CAP_bases',\\\n",
    "                                            '#GC_exons','#GC_bases','%bases_ovl','%UTR_bases_ovl','%exon_bases_ovl']])\n",
    "def merge_ranges(ls):\n",
    "    ls = np.copy(ls)\n",
    "    ls = ls[np.argsort(ls, axis = 0)[:,0]]\n",
    "    ovl_list =[]\n",
    "    for i in range(1,ls.shape[0]):\n",
    "        if (ls[i,0] < ls[i-1,1]):\n",
    "            ls[i,0] = np.min(ls[i-1:i+1,0])\n",
    "            ls[i,1] = np.max(ls[i-1:i+1,1])\n",
    "            ovl_list.append(i-1)\n",
    "    refined_list = np.delete(ls, ovl_list, 0)\n",
    "    return refined_list\n",
    "\n",
    "\n",
    "def remove_ranges(ls, ls2rm, with_indexes = False):\n",
    "\n",
    "    ls = np.copy(ls)\n",
    "    ls2rm = np.copy(ls2rm)\n",
    "    ls_ovl_indexes,ls2rm_ovl_indexes = search_overlap_partners(ls, ls2rm,return_indexes=True)\n",
    "    #print(search_overlap_partners(ls, ls2rm,return_indexes=True))\n",
    "    del_indexes = []\n",
    "    remain_indexes = []\n",
    "    for i in range(len(ls_ovl_indexes)):\n",
    "        ovl_region = np.array([np.max([ls[ls_ovl_indexes[i], 0], ls2rm[ls2rm_ovl_indexes[i],0]]),\\\n",
    "                               np.min([ls[ls_ovl_indexes[i], 1], ls2rm[ls2rm_ovl_indexes[i], 1]])])\n",
    "        if (ovl_region[1]-ovl_region[0])==(ls[ls_ovl_indexes[i], 1] - ls[ls_ovl_indexes[i], 0]):\n",
    "            del_indexes.append(i)\n",
    "            #print('UTR fl')\n",
    "        elif ovl_region[1] < ls[ls_ovl_indexes[i],1]:\n",
    "            remain_indexes.append(i)\n",
    "            ls[ls_ovl_indexes[i],0] = ovl_region[1]+1\n",
    "        elif ovl_region[0]>ls[ls_ovl_indexes[i],0]:\n",
    "            remain_indexes.append(i)\n",
    "            ls[ls_ovl_indexes[i],1] = ovl_region[0]-1\n",
    "        else:\n",
    "            raise ValueError('UTR found inside coding regions')\n",
    "    #print(ls_ovl_indexes[del_indexes])\n",
    "    ls = np.delete(ls, np.array(ls_ovl_indexes)[del_indexes], 0)\n",
    "    #print(ls)\n",
    "    if with_indexes:\n",
    "        return ls, remain_indexes\n",
    "    else:\n",
    "        return ls\n",
    "\n",
    "def mid_exons_search(gencode_gene_data):\n",
    "    #keep true internal exons intact\n",
    "    tx_list = unique_unsorted_elements(gencode_gene_data.transcript_id)\n",
    "    mid_exons = np.empty([1,2], dtype=int)\n",
    "    for tx_id in tx_list:\n",
    "        if gencode_gene_data[(gencode_gene_data.transcript_id == tx_id) &\\\n",
    "                    (gencode_gene_data.feature == 'UTR')].shape[0] > 0.:\n",
    "            exon_list = gencode_gene_data[(gencode_gene_data.transcript_id == tx_id) &\\\n",
    "                        (gencode_gene_data.feature == 'exon')]\n",
    "            exon_list = np.array(exon_list[['start','end']])\n",
    "            exon_list = exon_list[np.argsort(exon_list, axis = 0)[:,0]]\n",
    "            \n",
    "            mid_exon_point = np.average([np.min(exon_list[:,0]), np.max(exon_list[:,1])]) # sure fall at middle of exon seq\n",
    "            UTR_list = gencode_gene_data[(gencode_gene_data.transcript_id == tx_id) &\\\n",
    "                    (gencode_gene_data.feature == 'UTR')]\n",
    "            UTR_list = np.array(UTR_list[['start','end']])\n",
    "            UTR_list = UTR_list[np.argsort(UTR_list, axis=0)[:,0]]\n",
    "            # classify 3' and 5' UTRs\n",
    "            UTRs_5_prime = UTR_list[UTR_list[:,0]<mid_exon_point]\n",
    "            UTRs_3_prime = UTR_list[UTR_list[:,1]> mid_exon_point]\n",
    "            if UTRs_5_prime.shape[0]==0:\n",
    "                UTRs_5_prime = np.array([[np.inf,np.inf]])\n",
    "            if UTRs_3_prime.shape[0] == 0:\n",
    "                UTRs_3_prime = np.array([[-np.inf,-np.inf]])\n",
    "\n",
    "            mid_exon_bounds = np.array([np.max(UTRs_5_prime[:,1]), np.min(UTRs_3_prime[:,0])])\n",
    "            mid_exons = np.append(mid_exons, exon_list[(exon_list[:,0]>mid_exon_bounds[0]) &\\\n",
    "                                                        (exon_list[:,1]<mid_exon_bounds[1])], axis = 0)\n",
    "    mid_exons = np.delete(mid_exons,0, axis=0)\n",
    "    mid_exons = merge_ranges(mid_exons)\n",
    "    return mid_exons\n",
    "            \n",
    "        \n",
    "        \n",
    "def non_matching_exon_UTRs_search(exons, UTRs, *args):\n",
    "    #NME for non matching exon\n",
    "    #take the coordinates of ending exons\n",
    "    exons = np.copy(exons)\n",
    "    UTRs = np.copy(UTRs)\n",
    "    gencode_gene_data = pd.DataFrame.copy(args[0])\n",
    "    \n",
    "    final_exon_ends = [exons[0,1], exons[-1,0]]\n",
    "    #print(final_exon_ends)\n",
    "    otm_UTRs = np.append(UTRs[(UTRs[:,1]<=final_exon_ends[0])], UTRs[(UTRs[:,0]>=final_exon_ends[1])],axis=0) #outermost UTRs\n",
    "    #print(otm_UTRs)\n",
    "    #print(otm_UTRs)\n",
    "    mid_UTRs = np.array([UTRs[i] for i in range(UTRs.shape[0]) if UTRs[i,0] not in otm_UTRs[:,0]])\n",
    "    mid_exons = mid_exons_search(gencode_gene_data) #args_0 for gencode_gene_data\n",
    "    if mid_UTRs.shape[0]>0:\n",
    "        UTR_ME_indexes, check = search_overlap_partners(mid_UTRs, mid_exons, return_indexes= True)\n",
    "        NME_UTRs = np.delete(mid_UTRs, UTR_ME_indexes, axis= 0)\n",
    "        NME_UTRs = np.append(otm_UTRs, NME_UTRs, axis = 0)\n",
    "        NME_UTRs = merge_ranges(NME_UTRs)\n",
    "    else:\n",
    "        NME_UTRs = otm_UTRs\n",
    "    #print(NME_UTRs)\n",
    "    return NME_UTRs\n",
    "    \n",
    "def info_per_gene(gencode_table, capture_list, gene_name):\n",
    "    gencode_table = pd.DataFrame.copy(gencode_table)\n",
    "    capture_list = pd.DataFrame.copy(capture_list)\n",
    "    temp_table = temporary_table()\n",
    "    #deal with individual gene\n",
    "    gencode_gene_data = gencode_table[gencode_table.gene_name ==gene_name]\n",
    "    chr_name = unique_unsorted_elements(gencode_gene_data.seqname)[0]\n",
    "    gene_id = unique_unsorted_elements(gencode_gene_data.gene_id)[0]\n",
    "    \n",
    "    #after hg38 capture exons are no longer overlapping\n",
    "    #prep capture_list\n",
    "    capture_chr_data = pd.DataFrame.copy(capture_list[capture_list[0]==chr_name])\n",
    "    capture_chr_data.sort_values(1, axis=0, inplace= True)\n",
    "    capture_chr_data = np.array(capture_chr_data[[1,2]])\n",
    "    \n",
    "    #bases (exons+UTRs)\n",
    "    gencode_bases_data = gencode_gene_data[(gencode_gene_data.feature=='exon') | (gencode_gene_data.feature=='UTR')]\n",
    "    gencode_bases_data = np.array(gencode_bases_data[['start','end']])\n",
    "    gencode_bases_data = gencode_bases_data[np.argsort(gencode_bases_data, axis = 0)[:,0]]\n",
    "    #merging bases:\n",
    "    gencode_bases_data = merge_ranges(gencode_bases_data)\n",
    "    \n",
    "    #exons (before removing UTRs)\n",
    "    gencode_exons_data = gencode_gene_data[(gencode_gene_data.feature=='exon')]\n",
    "    gencode_exons_data = np.array(gencode_exons_data[['start','end']])\n",
    "    gencode_exons_data = gencode_exons_data[np.argsort(gencode_exons_data, axis = 0)[:,0]]\n",
    "    gencode_exons_data = merge_ranges(gencode_exons_data)\n",
    "    \n",
    "    #non-exon-matching UTRs (NEM UTRs)\n",
    "    gencode_UTRs_data = gencode_gene_data[gencode_gene_data.feature == 'UTR']\n",
    "    if gencode_UTRs_data.shape[0]>0:\n",
    "        #detect ovl UTR\n",
    "        gencode_UTRs_data.sort_values(by = 'start', axis=0, inplace= True)\n",
    "        gencode_UTRs_data = np.array(gencode_UTRs_data[['start','end']])\n",
    "        gencode_UTRs_data = merge_ranges(gencode_UTRs_data)\n",
    "        gencode_NEM_UTRs_data = non_matching_exon_UTRs_search(gencode_exons_data, gencode_UTRs_data,\\\n",
    "                                                             gencode_gene_data) \n",
    "        \n",
    "        #remove UTRs that does have any matches with exons to get exons only\n",
    "\n",
    "        gencode_exons_data = remove_ranges(gencode_exons_data, gencode_NEM_UTRs_data)\n",
    "        #% non-exon-matching UTRs:\n",
    "        UTR_ranges_ovl, num_cap_ovl_UTRs, ranges_cap_ovl_UTR = search_overlap_partners(capture_chr_data,\\\n",
    "                                                                                       gencode_NEM_UTRs_data)\n",
    "        total_UTR_bases = np.sum(gencode_NEM_UTRs_data[:,1] -gencode_NEM_UTRs_data[:,0] +1)\n",
    "        total_UTR_ovl_bases = np.sum(UTR_ranges_ovl)\n",
    "        percent_UTR_bases_ovl = total_UTR_ovl_bases/total_UTR_bases\n",
    "    else:\n",
    "        percent_UTR_bases_ovl = np.nan\n",
    "    \n",
    "    # % bases ovl:\n",
    "    bases_ranges_ovl, num_cap_ovl_bases, ranges_cap_ovl_bases = search_overlap_partners(capture_chr_data, gencode_bases_data)\n",
    "    total_GC_bases = np.sum(gencode_bases_data[:,1]- gencode_bases_data[:,0] +1) #not only those ovl\n",
    "    total_GC_ovl_bases = np.sum(bases_ranges_ovl)\n",
    "    \n",
    "    # % exons ovl:\n",
    "    exons_ranges_ovl, num_cap_ovl_exons, ranges_cap_ovl_exons = search_overlap_partners(capture_chr_data, gencode_exons_data)\n",
    "    total_GC_exons = np.sum(gencode_exons_data[:,1]- gencode_exons_data[:,0] +1) #not only those ovl\n",
    "    total_GC_ovl_exons = np.sum(exons_ranges_ovl)\n",
    "    temp_table['gene_name'] = [gene_name]\n",
    "    temp_table['gene_id'] = [gene_id]\n",
    "    temp_table['#TX'] = [gencode_gene_data[gencode_gene_data.feature == 'transcript'].shape[0]]\n",
    "    temp_table['#GC_exons'] = [gencode_exons_data.shape[0]]\n",
    "    temp_table['#GC_bases'] = [total_GC_bases]\n",
    "    temp_table['#CAP_exons'] = [num_cap_ovl_bases]\n",
    "    temp_table['%UTR_bases_ovl'] = [percent_UTR_bases_ovl]\n",
    "    temp_table['#CAP_bases'] = [np.sum(ranges_cap_ovl_exons)]\n",
    "    temp_table['%bases_ovl'] = [total_GC_ovl_bases/total_GC_bases]\n",
    "    temp_table['%exon_bases_ovl'] = [total_GC_ovl_exons/total_GC_exons]\n",
    "    temp_table['%UTR_bases_ovl'] = percent_UTR_bases_ovl\n",
    "    return temp_table\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>#TX</th>\n",
       "      <th>#CAP_exons</th>\n",
       "      <th>#CAP_bases</th>\n",
       "      <th>#GC_exons</th>\n",
       "      <th>#GC_bases</th>\n",
       "      <th>%bases_ovl</th>\n",
       "      <th>%UTR_bases_ovl</th>\n",
       "      <th>%exon_bases_ovl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMER1</td>\n",
       "      <td>ENSG00000184675.8</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3511</td>\n",
       "      <td>1</td>\n",
       "      <td>8443</td>\n",
       "      <td>0.011963</td>\n",
       "      <td>0.030171</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name            gene_id #TX #CAP_exons #CAP_bases #GC_exons #GC_bases  \\\n",
       "0     AMER1  ENSG00000184675.8   2          2       3511         1      8443   \n",
       "\n",
       "  %bases_ovl %UTR_bases_ovl %exon_bases_ovl  \n",
       "0   0.011963       0.030171             1.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_name ='AMER1'\n",
    "info_per_gene(gencode_protein_coding_gene, capture_data, gene_name = gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:34<00:00, 11.82s/it]\n"
     ]
    }
   ],
   "source": [
    "#multiprocessing:\n",
    "complete_gene_info = pd.DataFrame(columns=[['gene_name','gene_id','#TX','#CAP_exons','#CAP_bases',\\\n",
    "                                            '#GC_exons','#GC_bases','%bases_ovl','%UTR_bases_ovl','%exon_bases_ovl']])\n",
    "temp_gene_info = pd.DataFrame.copy(complete_gene_info) #temporary table used for updating\n",
    "gene_func_inputs = {'gencode_table':gencode_protein_coding_gene, \\\n",
    "                    'capture_list': capture_data, 'gene_name':gene_collection}\n",
    "def multi_gene_func(niter, **kwargs):\n",
    "    #print(kwargs['gene_name'])\n",
    "    kwargs['gene_name'] = kwargs['gene_name'][niter]\n",
    "    time.sleep(1)\n",
    "    return info_per_gene(**kwargs)\n",
    "mapfunc = partial(multi_gene_func,**gene_func_inputs)\n",
    "#pool = multiprocessing.Pool(5)\n",
    "#pool = multiprocessing.pool.ThreadPool(4)\n",
    "#outputs = []\n",
    "\n",
    "with multiprocessing.Pool(3) as p:\n",
    "      outputs = list(tqdm(p.imap(mapfunc, range(3)), total=3))\n",
    "#outputs = pool.map(mapfunc, range(len(gene_collection[:100])))\n",
    "complete_gene_info = pd.concat(outputs)\n",
    "complete_gene_info.reset_index(drop = True, inplace= True)\n",
    "\n",
    "#This step finished on terminal python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#complete_gene_info = pd.concat([complete_gene_info, temp_gene_info])\n",
    "complete_gene_info.reset_index(drop = True, inplace= True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19712, 10)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_gene_info = pd.read_csv('complete_genome_data.txt', sep =\"\\t\")\n",
    "complete_gene_info.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>#TX</th>\n",
       "      <th>#CAP_exons</th>\n",
       "      <th>#CAP_bases</th>\n",
       "      <th>#GC_exons</th>\n",
       "      <th>#GC_bases</th>\n",
       "      <th>%bases_ovl</th>\n",
       "      <th>%UTR_bases_ovl</th>\n",
       "      <th>%exon_bases_ovl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092.4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>941</td>\n",
       "      <td>1</td>\n",
       "      <td>918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FO538757.3</td>\n",
       "      <td>ENSG00000279928.1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>718</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>ENSG00000279457.2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>1982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000278566.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>961</td>\n",
       "      <td>1</td>\n",
       "      <td>939</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ENSG00000273547.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>951</td>\n",
       "      <td>1</td>\n",
       "      <td>939</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name            gene_id  #TX  #CAP_exons  #CAP_bases  #GC_exons  \\\n",
       "0       OR4F5  ENSG00000186092.4    1           1         941          1   \n",
       "1  FO538757.3  ENSG00000279928.1    1           0           0          3   \n",
       "2  FO538757.2  ENSG00000279457.2    3           0           0         14   \n",
       "3      OR4F29  ENSG00000278566.1    1           1         961          1   \n",
       "4      OR4F16  ENSG00000273547.1    1           1         951          1   \n",
       "\n",
       "   #GC_bases  %bases_ovl  %UTR_bases_ovl  %exon_bases_ovl  \n",
       "0        918         1.0             1.0              1.0  \n",
       "1        718         0.0             0.0              0.0  \n",
       "2       1982         0.0             0.0              0.0  \n",
       "3        939         1.0             1.0              1.0  \n",
       "4        939         1.0             1.0              1.0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1a6bbab0f0>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZEAAAEcCAYAAAAGD4lRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3Xtcj3f/B/BXxVdUKFpucihTTKlIoWJRyaHmHEvIKZs5TG5rjplzWU5zHGW3wi3HoYaMtvK7TWPdbM2kg5A2JelAp8/vD3fXfFfU91KKvZ6PR4+HPt/3dfX5fNuuV5/rc32vS00IIUBERCSDem13gIiIXl8MESIiko0hQkREsjFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiBARkWwMESIiko0hQkREsjFEiIhINoYIERHJVq+2O1BTHjzIQ2kpb1BMRFQV6upq0NXVUnm7NzZESksFQ4SIqIbxdBYREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyvbGfEyEielPs2hWOjIy8KtUaGGhh0qSRNdyjPzFEiIjquIyMPBgaTqlS7e3bX9Zwb5TxdBYREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyMUSIiEg2hggREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyMUSIiEg2hggREclW5RC5dOkSxowZAwsLC9jb22PZsmXIy/vzcY0xMTEYPnw4LCws0LdvXwQHB5fbx9WrV+Hl5QUrKyvY29sjKCgIRUVFSjUpKSmYNm0arK2tYWtriyVLliA3N/clhkhERDWlSiHy008/wdvbG/r6+ti6dSumT5+Or7/+GgsXLgQAXL58GdOmTYOxsTE2bdoENzc3BAQEYNeuXdI+UlNTMWHCBDRo0ADr16/HxIkTERISglWrVkk1Dx8+xPjx43H//n2sWbMGvr6+iIiIgK+vbzUPm4iIqkOVnrG+du1aWFpaYsOGDVBTU0OvXr1QWlqKkJAQFBQUYOPGjXjnnXcQGBgIAOjduzeKi4uxbds2eHl5QaFQYMeOHdDR0cGWLVugUCjQp08faGpqYvny5fDx8YGBgQHCwsKQk5ODo0ePQldXFwBgYGCAqVOnIj4+HhYWFjX3ThARkcoqnYlkZWUhLi4OY8aMgZqamtTu6emJqKgoqKurIy4uDi4uLkrb9e/fHzk5Obh8+TIAIDY2Fo6OjlAoFFKNq6srSkpKEBMTI9V0795dChAAsLe3h5aWFqKjo19upEREVO0qDZHffvsNQgg0adIEs2fPhqWlJbp164YlS5bg8ePHSEtLQ1FREYyMjJS2a9u2LQAgOTkZBQUFSE9PL1ejp6cHbW1tJCcnAwCSkpLK1WhoaMDQ0FCqISKiuqPS01lZWVkAAD8/Pzg7O2Pr1q24fv061q9fjydPnsDDwwMAoK2trbSdlpYWACA3NxePHj2qsKasrmzh/NGjR5XWEBFR3VFpiJRdPdW1a1csWbIEANCzZ08IIbBmzRqMGjUKAJROdT1LXV0dQojn1gghoK7+54SoKjVERFQ3VHpkLptR9O7dW6nd3t4eQghcvXoVAMrNFMq+19HRkWYXFc0m8vPzoaOjA+DpTKWimry8vApnKEREVLsqDZF27doBAAoLC5Xay2YohoaG0NDQwK1bt5ReL/veyMgIWlpaMDAwQGpqqlJNZmYmcnNzpXUQIyOjcjUlJSW4fft2ubUSIiKqfZWGSPv27dGqVStEREQotZ87dw716tWDlZUVrK2tcfr0aem0FQCcOnUKOjo6MDMzAwDY2dnh3LlzSmF06tQpaGhowMbGRqq5ePEisrOzpZqYmBjk5+ejV69eLzdSIiKqdhr+/v7+LypQU1ODvr4+QkJCkJKSAm1tbURGRmLz5s3w8vKCs7MzWrRogW3btuHmzZto2LAhjh49ii+//BIzZsyAra0tgKezjODgYMTFxaFJkyY4f/48AgMDMXLkSLi5uQEA3n77bezfvx9RUVFo1qwZLl++DH9/f9ja2sLHx0elgRUUFOKZTCMiem19//1PaNy4W5Vqc3Iuw8HBUuWfoaamhkaNFJUX/nU7Iap2qI2KisLmzZuRmJiIZs2awcPDAz4+PtKC95kzZ7Bx40YkJyfDwMAAnp6emDhxotI+4uLiEBAQgISEBOjq6mLIkCGYMWMG6tevL9X89ttvWLlyJa5cuQItLS04OTlh3rx5Kq+JZGbmorSUKUJEr7+VK3fD0HBKlWpv3/4S8+dPUPlnqKuroVkz1deeqxwirxuGCBG9KepyiPC6WSIiko0hQkREsjFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiBARkWwMESIiko0hQkREsjFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiBARkWwMESIiko0hQkREsjFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiBARkWwMESIiko0hQkREsjFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiBARkWwMESIiko0hQkREsjFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiBARkWwMESIiko0hQkREsjFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiBARkWyyQuSjjz6Cs7OzUltMTAyGDx8OCwsL9O3bF8HBweW2u3r1Kry8vGBlZQV7e3sEBQWhqKhIqSYlJQXTpk2DtbU1bG1tsWTJEuTm5srpJhER1TCVQ+TYsWM4c+aMUtvly5cxbdo0GBsbY9OmTXBzc0NAQAB27dol1aSmpmLChAlo0KAB1q9fj4kTJyIkJASrVq2Sah4+fIjx48fj/v37WLNmDXx9fREREQFfX9+XGCIREdWUeqoUZ2RkYMWKFWjRooVS+8aNG/HOO+8gMDAQANC7d28UFxdj27Zt8PLygkKhwI4dO6Cjo4MtW7ZAoVCgT58+0NTUxPLly+Hj4wMDAwOEhYUhJycHR48eha6uLgDAwMAAU6dORXx8PCwsLKpp2EREVB1UmoksXLgQdnZ26Nmzp9T25MkTxMXFwcXFRam2f//+yMnJweXLlwEAsbGxcHR0hEKhkGpcXV1RUlKCmJgYqaZ79+5SgACAvb09tLS0EB0drfroiIioRlU5RMLDw/Hzzz9j0aJFSu1paWkoKiqCkZGRUnvbtm0BAMnJySgoKEB6enq5Gj09PWhrayM5ORkAkJSUVK5GQ0MDhoaGUg0REdUdVTqddefOHaxatQqrVq2Cnp6e0muPHj0CAGhrayu1a2lpAQByc3OfW1NWV7Zw/ujRo0priIio7qh0JiKEwPz589GnTx/079+/wtcBQE1NreIfoK7+whohBNTV/+xGVWqIiKhuqHQmEhYWhuvXr+P48eMoLi4G8GdwFBcXQ0dHBwDKzRTKvtfR0ZFmFxXNJvLz86V9aGtrV1iTl5eHVq1aVXlQRET0alQaIqdOncKDBw9gb29f7rXOnTvD398fGhoauHXrltJrZd8bGRlBS0sLBgYGSE1NVarJzMxEbm6utA5iZGRUrqakpAS3b9+ucBZERES1q9JzREuXLsXBgweVvhwdHdGiRQscPHgQrq6usLa2xunTp6UZCvA0fHR0dGBmZgYAsLOzw7lz51BYWKhUo6GhARsbG6nm4sWLyM7OlmpiYmKQn5+PXr16VdugiYioelQ6EzE2Ni7X1rRpUygUCpibmwMAPvjgA3h7e+Pjjz/G0KFDceXKFezatQu+vr5o2LAhAGDy5Mk4efIkpk6divHjxyMlJQVBQUEYNWoUWrZsCQB4//33ERoaigkTJmD69OnIzs5GYGAgevfuja5du1bnuImIqBpUy2p1z549sWnTJty8eRPTp0/H8ePHMW/ePEyZMkWqad++PYKDg5Gfn4+ZM2ciJCQE3t7eWLBggVSjp6eHf/3rX2jatCnmzp2LdevWwdXVFevWrauObhIRUTVTE8+eg3qDZGbmorT0jRwaEf3NrFy5G4aGUyovBHD79peYP3+Cyj9DXV0NzZqV/4hFpdupvAUREdH/MESIiEg2hggREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyMUSIiEg2hggREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyMUSIiEg2hggREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyMUSIiEg2hggREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyMUSIiEg2hggREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSjSFCRESyMUSIiEg2hggREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKSrUohUlpain379sHNzQ1WVlZwcnLCqlWrkJubK9VcvXoVXl5esLKygr29PYKCglBUVKS0n5SUFEybNg3W1tawtbXFkiVLlPYBAPfv34evry9sbW3RrVs3zJkzB3/88Uc1DJWIiKpbvaoU7dy5E+vXr8ekSZPQs2dPJCcnY+PGjUhMTMSuXbuQmpqKCRMmwMrKCuvXr8fNmzexbt065ObmYvHixQCAhw8fYvz48dDX18eaNWuQmZmJwMBA3Lt3D9u3bwcAFBcXY9KkScjPz4e/vz+Ki4vx+eefY/LkyTh06BDq1atSd4mI6BWp9KgshMDOnTvh4eEBX19fAECvXr2gq6uLjz/+GAkJCQgNDYWOjg62bNkChUKBPn36QFNTE8uXL4ePjw8MDAwQFhaGnJwcHD16FLq6ugAAAwMDTJ06FfHx8bCwsMDJkyfx66+/IiIiAu3btwcAdOrUCYMHD8bp06cxcODAGnwriIhIVZWezsrLy4O7uzsGDx6s1G5sbAwAuHXrFmJjY+Ho6AiFQiG97urqipKSEsTExAAAYmNj0b17dylAAMDe3h5aWlqIjo6Wat5++20pQABI35fVEBFR3VHpTERbWxsLFy4s1x4VFQUAaN++PdLT02FkZKT0up6eHrS1tZGcnAwASEpKgru7u1KNhoYGDA0NlWr+uh8AaNOmjVRDRER1h6yrs+Lj47Fjxw44OTmhcePGAJ6GzV9paWlJC+ePHj2qlhoiIqo7VA6RH3/8EZMnT4ahoSGWL18OIQQAQE1NrVytEALq6n/+iOqqISKiukGlI3NERAS8vb3xj3/8A7t374aurq40c6hoppCfnw8dHR0AT2cqFdXk5eVJ+6hKDRER1R1VDpGQkBDMmTMHlpaWCAsLw1tvvQXg6akmAwMDpKamKtVnZmYiNzdXWuMwMjIqV1NSUoLbt2+/sAZ4unhf0VoJERHVriqFSHh4OFavXo0BAwZg586d0uyijJ2dHc6dO4fCwkKp7dSpU9DQ0ICNjY1Uc/HiRWRnZ0s1MTExyM/PR69evQA8vVrrxo0bSEpKkmoSExORlJQk1RARUd2hJsoWNZ4jMzMT/fr1g56eHgICAsp94K9NmzZ48OABhg4diq5du2L8+PFISUlBUFAQhg8fDn9/fwBAVlYWBg4ciBYtWmD69OnIzs5GYGAgLCws8OWXXwIACgsL4e7ujsLCQvj6+kIIgc8//xza2to4cuSISh82zMzMRWnpC4dGRPRaWLlyNwwNp1Sp9vbtLzF//gSVf4a6uhqaNVN92aDSEDl69Cg++eST574eEBCA9957D3FxcQgICEBCQgJ0dXUxZMgQzJgxA/Xr15dqf/vtN6xcuRJXrlyBlpYWnJycMG/ePKX1jvT0dKxYsQKxsbFQKBSws7ODn5+fdPqsqhgiRPSmeK1D5HXFECGiN0VdDhFeN0tERLIxRIiISDaGCBERycYQISIi2RgiREQkG0OEiIhkY4gQEZFsDBEiIpKNIUJERLIxRIiISDaGCBERycYQISIi2RgiREQkG0OEiIhkY4gQEZFsDBEiIpKNIUJERLIxRIiISDaGCBERycYQISIi2RgiREQkG0OEiIhkY4gQEZFsDBEiIpKNIUJERLIxRIiISDaGCBERycYQISIi2RgiREQkG0OEiIhkY4gQEZFsDBEiIpKtXm13gIiouu3aFY6MjLxK65KSrsPY2LRK+zQw0MKkSSNftmtvHIYIEb1xMjLyYGg4pdK62NgZ6N278joAOHFiZpWC6e8WNgwRIqpVNTFr+OmnBBgavmzPlOXliSoFU1XDBqj6mGpiPNWFIUL0N1fVgzhQ9YOeqgf8wYODKq1TZdYQGzujSnU1oaphA1R9TLU5nsowRIjqgKoeyFU5VVLVfVb1IA6odtB7HQ749PIYIkR1QFXP4atyqkSVv/CJ5GKIENUQVU4TVfWct6qnSohqGkOE3mg1sWirymKoKqeJiF5HDBF6o9XEpZ5vwmIoUXVhiFCdUROLy0RUsxgiVOOq+yohVReX6+r19URvAoYIyaLqonF1XiXExWWiuoMhQkpq6rMFRPRmYohU4E07N1+bswYierMxRCpQ1St6bt/+8hX05uVVdTwAw4GIVMMQeY2pcuqJi8tEVBPqZIicOHECW7duRVpaGlq1agUfHx8MGTKktrtV56jyGQgioppQ50IkMjISc+fOxbhx4+Dg4ICoqCh88skn0NTUhKura213r8bVxK0yiIhqSp0LkaCgIAwYMADz588HADg4OODhw4fYsGHDax0ivOqJiN5EdSpE0tLScOvWLcyZM0epvX///oiMjERaWhpat25dS70r76efrmLlyt1VrOVVT0T05qlTIZKUlAQAMDIyUmpv27YtACA5ObnKIaKuria7H7q6OmjcuPLtGzfWwzvvTK3SPpOSllRpn//4R7Mq1alS+6bt800bD/fJfVbnPnV1dWQd/+QeM9WEEELWljXgxIkT8PX1xdmzZ2H4zMn+1NRUuLi4YN26dRg4cGAt9pCIiJ6lXtsdeFZZnqmpqVXYrq5ep7pLRPS3V6eOyjo6OgCA3Nxcpfa8vDyl14mIqG6oUyFSthZy69YtpfbU1FSl14mIqG6oUyHStm1bGBoa4ptvvlFqP336NNq1a4eWLVvWUs+IiKgiderqLACYPn06Pv30UzRp0gTvvvsuvv32W0RGRmLdunW13TUiIvqLOnV1Vpn9+/cjODgY6enpaN26NaZOncrbnhAR1UF1MkSIiOj1UKfWRIiI6PXCECEiItn+diFy4sQJDBo0CF26dMGAAQNw9OjRF9bn5eVh6dKlsLOzg5WVFaZMmYKUlJRX09lqouqY//jjDyxcuBCOjo6wsrLCsGHDEBkZ+Yp6Wz1UHfOz0tPT0a1bN2zZsqUGe1i9VB1vaWkptm7din79+qFLly5wc3PDyZMnX1Fvq4eqY87KysKnn34Ke3t72NjYwMfH57X7f7lMQkICOnfujHv37r2w7pUcv8TfSEREhDA1NRUrVqwQ3333nVi8eLEwMTERkZGRz91mypQpokePHuLw4cPi1KlTws3NTTg4OIicnJxX2HP5VB3zkydPhLu7u3B0dBSHDx8WMTExYtGiRcLExEQcP378FfdeHjm/5zKlpaViwoQJwsTERGzevPkV9PblyRnvsmXLhJmZmQgODhYXLlwQCxYsEKampuL8+fOvsOfyqTrm0tJSMXr0aNGrVy9x5MgRce7cOTF06FDh4OAgsrOzX3HvX87NmzeFg4ODMDExEenp6S+sfRXHr79ViDg5OYnZs2crtc2aNUu4urpWWH/p0iVhYmIioqOjpbbMzExhaWkptm/fXqN9rS6qjvnMmTPCxMRExMfHK7VPmjRJuLu711g/q5OqY35WaGio6N2792sVIqqONzU1VXTs2FEcOHBAqd3T01MsW7asxvpZnVQdc1JSkjAxMRFHjhyR2m7duiVMTEzE4cOHa7Sv1aWoqEiEhoYKKysrYWNjU2mIvKrj19/mdFbZbeZdXFyU2vv374+kpCSkpaWV2yY2NhZaWlqws7OT2vT09NC9e3d89913Nd7nlyVnzFpaWvDw8IC5ublSu7Gxcbk7CdRFcsb87LZr167FsmXLarqb1UbOeKOioqCpqVnusvnQ0FAsXLiwRvtbHeSM+cmTJwCe/vddpkmTJgCA7OzsGuxt9fnxxx+xdu1aTJw4EXPnzq20/lUdv/42IVKV28xXtE3btm2hoaGh1N6mTZsK6+saOWPu2bMnPvvsM6WbYBYVFSE6OhodOnSowd5WDzljBp6uEfj5+WHAgAHo3bt3zXayGskZ7/Xr12FkZIQLFy7A3d0d77zzDlxcXBAREVHzHa4GcsbcsWNH2NraYvPmzbh58yaysrKwfPlyNGrUCE5OTjXf6WrQvn17REVF4aOPPip3TKrIqzp+1blPrNeUR48eAQC0tbWV2sv+MvnrTR/L2v5aX7ZNRfV1jZwxV2Tt2rVISUnB5s2bq7eDNUDumL/66iukpaVh27ZtNdvBaiZnvFlZWUhPT8f8+fMxa9YsGBoaIjw8HB9//DH09PTQo0ePmu/4S5D7O/b398fkyZOlx0koFAps3ry5Tj3o7kWaN2+uUv2rOn79bUJEyLjNvHjB5zBfh9vSyxnzX+sCAwOxe/duTJo06bX4i03OmJOSkrB+/Xps3LjxtbtTtJzxFhUVISsrC9u2bYOjoyOApzPQpKQkfPHFF3U+ROSM+ebNmxg9ejTatGmD+fPnQ1NTEwcOHMDMmTOxc+dOWFtb13zHX7FXdfyq+0fCaiLnNvPa2trS63/dpqKEr2te5tb6hYWF8PX1xa5duzBp0iTMmzev5jpajVQdc0lJCfz8/ODq6go7OzsUFxejuLgYwNNTXGX/rqvk/I61tLSgoaGhdK5cTU0NvXr1wvXr12uwt9VDzph3794NAAgODoaTkxPs7e2xYcMGdOrUCStXrqzZDteSV3X8+tuEiJzbzBsZGSEtLa1coqempr4Wt6WXe2v93NxceHt7IzIyEvPnz39tAgRQfczp6emIj4/H0aNH0blzZ+kLADZt2iT9u66S8ztu27ZthQFZVFRU7q/7ukjOmO/evYv27dtLi+nA0+Ds1q0bEhMTa7C3tedVHb/+NiEi5zbz9vb2yMnJwYULF6S2rKwsxMXFoVevXjXe55clZ8wlJSX44IMPEB8fj6CgIIwfP/5VdbdaqDrmt956CwcPHiz3BQBjxoyR/l1XyfkdOzg4QAih9AHS4uJifP/99+jWrVuN9/llyRmzkZERbty4gYcPHyq1x8fHo1WrVjXa39ryqo5fGv7+/v7Vtrc6TkdHB1u3bsWDBw+gpqaGkJAQHDlyBEuWLEGHDh2QlZWF69evQ1tbGwqFAq1atcIPP/yAvXv3omnTprh79y7mz58PIQRWrlwJTU3N2h5SpVQd8759+7B3716MHDkSNjY2uHfvnvSVkZGBFi1a1PaQKqXKmBs2bAgDA4NyX1988QX69euHwYMH1/ZwKqXq77h169a4evUq9u/fDy0tLeTk5CAgIADXrl1DQEAA3nrrrdoeUqVUHbOxsTHCw8MRHR2Npk2b4s6dO9i4cSPOnTuHxYsXvxZXHj4rISEBZ8+ehbe3t3RqqtaOX9X2iZPXxL59+4Szs7MwMzMTAwYMUPrw0aFDh4SJiYn4z3/+I7VlZ2cLPz8/YW1tLbp27SqmTJkibt68WRtdl02VMXt5eQkTE5MKvzp16lRbQ1CZqr/nv3qdPmwohOrjLSgoEKtXrxb29vbC3NxceHh4iIsXL9ZG12VTdcyJiYnCx8dHWFlZiW7duokxY8aI2NjY2uj6Sysb37MfNqyt4xdvBU9ERLL9bdZEiIio+jFEiIhINoYIERHJxhAhIiLZGCJERCQbQ4SIiGRjiFA5eXl5mDlzJiwsLODu7o64uLhyNXv37oWLiwtKSkpqoYc1r2/fvvDy8qrtbryx/Pz8YGpqWtvdoGrAEKFytm/fjgsXLmD27Nlo2bIlPvzwQ+Tk5EivFxYWYvv27fjwww+r9FwDInpzMUSonIiICIwZMwbe3t5Yu3Yt8vPzER0dLb3+73//G5qamnBzc6vFXhJRXcAQoXIyMjJgaGgI4OntpJs2bYp79+4BePqY0R07dmD69OmchRARQ4TK09XVlZ4eV1paitzcXOjq6gIA9u3bB21tbVk3Jrxy5Qq8vb1hZWUFKysrTJw4Ef/973+l17/77juYmppi1qxZStstWrQIpqamSs+FjouLw4QJE6R9jRs3DpcuXVLarm/fvli8eDGOHTuGQYMGwdzcHC4uLggLC6tyn8PDw9GvXz+Ym5tj5MiR+P7775VeF0Jg3759GDFiBKysrGBubg5XV1fs2LFD6RbcDx8+hJ+fH959912YmZnByckJn3/+ufTs7zKJiYmYPn06rK2tYWFhgdGjR5f7mYWFhVixYgX69esHMzMz9OnTB0uXLi13h9qK3LlzB//85z/Ro0cPmJubw93dHQcOHJBej4iIgKmpKU6fPl1uWy8vL6UHk927dw/z5s2T9jVkyBB8/fXXStuUPaslLCwM3bt3f+Hzvf/v//4PkydPhq2tLTp37gwHBwcsXrxY6VSqn58fnJ2dceXKFQwbNgxdunSBq6sr9u3bV+nYqWb8re7iS1Vz7do1fP/997C1tcWxY8cQExODefPmQVNTE7Nnz8acOXNUXhSNjY3FxIkT0bhxY4wbNw5du3bFDz/8gN27d8PGxgYtW7ZE27ZtcefOHZw4cQKWlpZo27YtYmNjsXLlSowePRoTJkwAAJw9exZTpkyBhoYGxo0bh+7du+PixYvYs2cPOnbsCGNjYwBPH3mbmpqKs2fPYtiwYejfvz+uX7+OQ4cOoUuXLmjXrt1z+/vVV1/h1q1biI6OxogRI9C3b1/8+OOPCAsLg5WVlfRI1fXr12Pt2rWws7PDqFGjYG1tjcTERBw/fhyGhobo1KkTAOCDDz7ApUuX8P7772PQoEHQ0NBAaGgoMjMz0bdvXwBPn30+ZswYPH78GOPGjYOdnR1+/fVXBAcHw9jYWLrT7JIlS3D06FGMHDkS7u7u0NXVxf79+3Ht2jW89957zx1TWloaRo4ciRs3bmDMmDFwcXHBnTt3EBoaioIOZx6XAAAMcUlEQVSCAtjZ2aF169bYs2cPioqK4OrqKm37+++/Y+XKlXj//ffRo0cPZGRkYOTIkUhLS4Onpyf69euH9PR0BAcHo2HDhujatSsAICoqCleuXMFPP/2EDz/8EO3atcOgQYMQExODX3/9FTNmzAAAxMTEYPLkydDX18fYsWPRu3dvFBYW4vjx48jMzJTCKyoqCgkJCTh27BgsLS0xYsQI3Lt3D2FhYVAoFG/kEwrrvGq9nSO9EW7duiWcnJyEiYmJ6NixowgJCRFCCLFz504xcOBAUVJSotL+SkpKRL9+/cTo0aNFcXGx1J6XlyecnZ3Fe++9J7VlZ2cLOzs74ezsLLKyssS7774rnJycRF5enhBCiKKiItG7d2/Rp08f8ejRI2m7hw8fCgcHB+Hg4CAKCwuFEEI4OjoKU1NTkZCQINX9/vvvwtTUVMyZM+eFfXZ0dBQmJibi/PnzUtuDBw+EjY2NGDp0qBBCiMLCQtG1a1fx8ccfK2376NEjYWZmJnx8fIQQQty/f1+YmJiInTt3KtX5+fmJ8ePHS9+PHTtWaaxl433//fdFr169xJMnT4QQQnTp0kUsXbpUaV/r1q0Tw4YNE7m5uc8d0+zZs0XHjh3FtWvXpLaSkhLh4+MjTE1NxW+//SaEEGLevHnC0tJSFBQUSHVfffWVMDExEYmJiUIIIT755BNhY2MjMjIylH7GnDlzhJmZmbh//75UZ2JiIg4dOqRUV9ZeZtKkScLR0VEaY5lRo0YJKyurctstX75caisuLhaenp6iS5cuIjs7+7njp5rB01lUTuvWrXHy5EkcOnQI3333HSZMmID8/Hzs3LkT06dPh7q6Oo4ePQpXV1c4Ojpi06ZNKC0tfe7+fvnlF6SlpcHJyQkPHz5EVlYWsrKy8PjxYzg6OiIhIUFac2nSpAmWLl2K1NRUjBw5Evfu3cOaNWvQqFEjaV/37t2Dp6en0iM+GzdujLFjxyIjIwPXrl2T2o2MjNCxY0fpe319fTRv3hz379+v9H3o0KED+vTpI33ftGlTuLm54eeff8Yff/yB+vXr48KFC/jss8+Utnvw4AG0tbWRn58P4OmzLxo1aoS9e/fi1KlTUvuqVaukx7Y+ePAAP/zwA/r06YPHjx9L71FOTg6cnZ1x//59XL16FQDQokULRERE4PDhw9KpntmzZ+PQoUPQ0tKqcCwlJSU4f/487O3tlZ7WqK6ujmnTpkEIgW+//RYA4ObmVu5iioiICHTu3Bnt27dHaWkpoqKiYG1tjXr16kl9zcrKgouLCwoLCxEbG6v08+3t7V/4Xm/fvh2HDh2CQqF47vv4LB8fH+nfZTPSx48fKz2AiV6NerXdAaqbFAoFzMzMpO9DQ0PRrFkzDBgwADdu3ICfnx8WLVqENm3awNfXFwYGBhg1alSF+yp7jGlAQAACAgIqrElPT5ceeNWvXz+4uLjg9OnTGDNmjHRqBABu374NoOJHoJadxrp79y6srKwAAHp6ehWO7UWh99f9PatNmzYAnq4t6Ovro379+jh//jzOnj2L5ORkpKamSmsT4n9rIgqFAp999hkWLVqEmTNnQqFQwMbGBi4uLhgyZAgaNGiAtLQ0AMCePXuwZ8+e575HAODv74/Zs2fj008/xaJFi2BpaQlnZ2cMHz68wueLA08PyPn5+RW+b+3bt5fGBAA9e/ZE8+bN8c0336B///64e/cufvrpJ3zyySfSvh49eoSoqChERUW9sK9lmjVrVmFdGQ0NDaSlpWHDhg1ITEzErVu3kJGRUWFt06ZN0bx5c6W2tm3bKo2BXh2GCFUqLy8PwcHB8Pf3h5qaGr755hu0adMGnp6eAABXV1ecPHnyuSFSdsCeNWsWLC0tK6x59oCdn5+PX375BQDw/fffIz8/X5qJiBc8/qbstfr160tt6urVO9kuG4u6ujqEEPjnP/+JEydOoFu3brCysoKHhwe6d+9e7rHCbm5ucHBwQFRUFKKjo3HhwgXExMRg7969CA8Plz606enpqbR4/ay3334bwNOD/Llz56Sv2NhYaVZz+PDhCoPzRe9b2ZjKZgEaGhoYMGAADh06hMePHyMyMhJqamoYOHAgAEh97d+/P0aPHl3hPsvWjMpUdiXf/v37sWTJEhgZGcHa2houLi6wsLDAnj17cPz4caXaZ3+/fx0Drxh89RgiVKk9e/ZAX18f/fv3BwDcv39f6UDVtGnTcldGPavsGdaNGjUq92zn//73v3j48KHSozqDgoJw584dzJs3D4GBgQgKCsLChQuV9pWUlFTu5yQnJwNAtT3Ct6K/alNTUwE8PUjGxcXhxIkT+PDDD5WuKCsuLkZ2drZ0IM3Ly0NCQgI6dOiAESNGYMSIESgsLERgYCD+9a9/ISYmRpr1aWholHuPEhMTcfv2bTRs2BCFhYVISEhAixYtMGjQIAwaNAilpaUICQlBQEAATp48WeEn7fX09NCoUaMqv2/u7u7Ys2cPYmNjERkZCRsbGxgYGEj7atiwIYqLi8v19e7du/jll1/QsGHDyt/g/3ny5AlWr14NW1tbBAcHo169Pw9LGzZsKFd///595OXlKZ26S0lJAfDnjIReHa6J0Avl5uYiJCQEM2bMgJqaGoCn6wrp6enSX7e3b9+WDjAVMTMzg76+Pvbs2YO8vDylfZedlin7C7LsCqhRo0Zh0qRJGD58OEJDQ6Vbr3Tu3Bn6+vrYt28fcnNzlfa1d+9e6OvrK52Gexk///yzNCMCnh68vv76a1hbW0NXVxfZ2dkA/pwhlDlw4AAKCgpQXFwMALhx4wY8PT1x8OBBqUahUOCdd94B8DQ43nrrLZiZmeHIkSNKp3GKioowf/58zJw5UwonDw8PbN++XapRV1eHubm59O+KaGhowMHBAbGxsfj555+ldiEEvvzyS6ipqeHdd9+V2suuXjtw4ACuXr2q9MHSevXqoXfv3oiOjsavv/6q9HNWr16N6dOn48GDBy94Z5U9fvwYBQUFaNeunVKAJCQk4IcffgAA6b0s6/Ozl2kXFxfjq6++go6ODnr27Fnln0vVgzMReqHdu3ejRYsWcHZ2ltqcnZ3xxRdfYMGCBWjdujVOnz6NJUuWPHcf9evXx6JFizB79mwMGzYMI0aMQIMGDRAeHo67d+9i7dq1qFevHp48eYIFCxZAT08Pc+fOBQDMnTsXUVFRWLBgAY4dOwZNTU1pX8OHD8eIESMAAAcPHsTvv/+OjRs3VtsprCZNmmDSpEnw9vaGhoYGwsLCUFxcjE8//RQAYGVlBW1tbaxatQp3795F48aNcfHiRURERKBBgwZSYFpYWMDa2hrr1q1Deno6TE1NkZ6ejtDQUBgbG0sHvoULF2L8+PEYPnw4xowZg6ZNm+LkyZOIj4+Hr6+v9FkdNzc37N27FwUFBbCyskJ2djZCQ0PRvHlzDBgw4LnjmTt3Li5evAgvLy94eXlBX18fZ86cwX/+8x94e3uXC8PBgwfjiy++gEKhkGahf92Xp6cnPD090bJlS5w/fx7nzp2Dh4eHdDlyVd9nCwsLHD58GNra2jAyMsKNGzcQHh4u/S7z8vLQpEkTaZstW7bgzp076NChAyIjI3HlyhWsWLFCpRkQVZNauy6M6rycnBxhbW0tzpw5U+61gwcPCkdHR9GjRw8RFBRUpct+L1y4ILy8vISlpaXo2rWr8PDwEN9++630emBgoDAxMRHHjh1T2i48PFyYmJiI1atXK+1r7NixwsLCQnTr1k1MnDhRXLp0SWk7R0dHMXbs2HL9eF77X2t8fX3F7t27hYODgzA3Nxdjx44V8fHxSnVxcXFi9OjRwtLSUtjY2AgPDw9x8uRJsWrVKtGpUyfxxx9/CCGeXh68bNky0bdvX2FmZibs7OzEggULxO+//660v2vXrgkfHx/RrVs3YWFhIYYMGSIOHz6sVFNQUCA2bNggXFxchLm5ubCxsRGzZs0SKSkpLxyTEEKkpKSI2bNnCxsbG9GlSxcxdOhQER4eXmFtcnKyMDExER999NFz9zVnzhxha2srzM3NxcCBA0VISIjSZdx/vZT3ee13794VM2bMEDY2NsLS0lIMGjRIbNmyRZw6dUqYmJiIb775Rmm7uLg4MXjwYGFubi6GDh0qTp8+XenYqWaoCfGCFTciojrEz88PR44cwfXr12u7K/Q/XBMhIiLZGCJERCQbQ4SIiGTjmggREcnGmQgREcnGECEiItkYIkREJBtDhIiIZGOIEBGRbAwRIiKS7f8B8lfNYbUcnfYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.set(style=\"dark\", color_codes=True, font_scale=1.5)\n",
    "sns.distplot(complete_gene_info['%exon_bases_ovl'].dropna(), hist=True, kde=False, color = 'blue',\\\n",
    "             hist_kws={'edgecolor':'black'}, norm_hist= False, axlabel='% exon bases overlap')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "remove 2-end exon portions/ exons that are UTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gencode_protein_coding_gene_rm_UTRexons.loc[55].seqname.fill('chrDMM')\n",
    "gencode_protein_coding_gene_rm_UTRexons.loc[55]=\\\n",
    "gencode_protein_coding_gene_rm_UTRexons.loc[55].replace('chr1','chrDMM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make tracker for testing genes\n",
    "gencode_protein_coding_gene[(gencode_protein_coding_gene.gene_name=='EPAS1')\\\n",
    "                            & (gencode_protein_coding_gene.feature=='exon')]\\\n",
    "[['seqname', 'start', 'end']].to_csv('sample_EPAS1_exon_track.txt', index=None, sep='\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_gene_info.to_csv('complete_genome_data.txt', index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
